---
title: "Bensidoun et al. 2021 AP-RNA-seq analysis"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = F, message = F, include = T, eval = T, warning = F,
                      cache = F)
```

```{r libraries}
library(dplyr)
library(readr)
library(tibble)
library(tidyr)
library(ggplot2)
library(DESeq2)
library(UpSetR)
library(xlsx)
library(rtracklayer)
library(ggpmisc)
library(ggpubr)
library(broom)
library(kableExtra)
library(clusterProfiler)
```

```{r helper_functions}
orf_to_common <- function(keys){
  commons <- AnnotationDbi::select(org.Sc.sgd.db::org.Sc.sgd.db,
                                   keys = keys,
                                   columns=c("COMMON"),
                                   keytype="ORF")
  commons <- commons[match(unique(commons$ORF), commons$ORF), ]
  commons <- commons[ , -2]
  commons$COMMON <- ifelse(is.na(commons$COMMON), commons$ORF, commons$COMMON)
  commons <- unique(commons[ , ])
  commons <- commons$COMMON
  return(commons)
}

orf_to_description <- function(keys){
  description <- AnnotationDbi::select(org.Sc.sgd.db::org.Sc.sgd.db,
                                       keys = keys,
                                       columns=c("DESCRIPTION"),
                                       keytype="ORF")
  description <- description[match(unique(description$ORF), description$ORF), ]
  description <- description$DESCRIPTION
  return(description)
}
```

```{r diffex}
logfc_threshold = 1

info <- read_csv("inputs/info.csv")
info$condition<- factor(info$condition, levels = c("input", "basket", "basketless", 
                                                   "total", "igg", "gbp"))
counts <- read_tsv("outputs/counts/raw_counts.tsv") %>%
  filter(!grepl("__", gene)) %>%
  select(!starts_with("AP_D")) %>%
  column_to_rownames("gene")

dds <- DESeqDataSetFromMatrix(counts,
                              colData = info,
                              design = ~ condition)
ds <- DESeq(dds, test="Wald")
#resultsNames(ds)
# log2 fold change (MLE): condition total vs input
# results are log2(total/input)
# log2FC values are expressed as that change in gene expression in total compared
# to that in input.

res_basket <- results(ds, contrast = c("condition", "basket", "input"), alpha = .05)
res_basket <- res_basket %>%
  as.data.frame() %>%
  rownames_to_column("gene") %>%
  filter(padj < 0.05) %>%
  filter(abs(log2FoldChange) > logfc_threshold) %>%
  mutate(common = orf_to_common(gene)) %>%
  mutate(description = orf_to_description(gene))

res_basketless <- results(ds, contrast = c("condition", "basketless", "input"), alpha = .05)
res_basketless <- res_basketless %>%
  as.data.frame() %>%
  rownames_to_column("gene") %>%
  filter(padj < 0.05) %>%
  filter(abs(log2FoldChange) > logfc_threshold) %>%
  mutate(common = orf_to_common(gene)) %>%
  mutate(description = orf_to_description(gene))

res_total <- results(ds, contrast = c("condition", "total", "input"), alpha = .05)
res_total <- res_total %>%
  as.data.frame() %>%
  rownames_to_column("gene") %>%
  filter(padj < 0.05) %>%
  filter(abs(log2FoldChange) > logfc_threshold)  %>%
  mutate(common = orf_to_common(gene)) %>%
  mutate(description = orf_to_description(gene))

res_gbp <- results(ds, contrast = c("condition", "gbp", "input"), alpha = .05)
res_gbp <- res_gbp %>%
  as.data.frame() %>%
  rownames_to_column("gene") %>%
  filter(padj < 0.05) %>%
  filter(abs(log2FoldChange) > logfc_threshold)  %>%
  mutate(common = orf_to_common(gene)) %>%
  mutate(description = orf_to_description(gene))

res_igg <- results(ds, contrast = c("condition", "igg", "input"), alpha = .05)
res_igg <- res_igg %>%
  as.data.frame() %>%
  rownames_to_column("gene") %>%
  filter(padj < 0.05) %>%
  filter(abs(log2FoldChange) > logfc_threshold)  %>%
  mutate(common = orf_to_common(gene)) %>%
  mutate(description = orf_to_description(gene))

# write.xlsx(res_total, file="outputs/xlsx/deseq2_results_background.xlsx", sheetName="res_total", row.names=FALSE)
# write.xlsx(res_basket, file="outputs/xlsx/deseq2_results_background.xlsx", sheetName="res_basket", row.names=FALSE, append = T)
# write.xlsx(res_basketless, file="outputs/xlsx/deseq2_results_background.xlsx", sheetName="res_basketless", row.names=FALSE, append = T)
# write.xlsx(res_igg, file="outputs/xlsx/deseq2_results_background.xlsx", sheetName="res_igg", row.names=FALSE, append = T)
# write.xlsx(res_gbp, file="outputs/xlsx/deseq2_results_background.xlsx", sheetName="res_gbp", row.names=FALSE, append = T)
```

```{r remove_specific_background}
# remove specific background from differentially expressed genes
remove_igg_background <- function(res, res_igg){
  res_igg_up <- res_igg %>%
    filter(log2FoldChange > logfc_threshold) %>%
    filter(padj < 0.05)
  
  res_igg_down <- res_igg %>%
    filter(log2FoldChange < -logfc_threshold) %>%
    filter(padj < 0.05)
  
  background_up <- c(res_igg_up$gene)
  background_down <- c(res_igg_down$gene)
  
  res_up <- res %>%
    filter(log2FoldChange > logfc_threshold) %>%
    filter(!gene %in% background_up)
  
  res_down <- res %>%
    filter(log2FoldChange < -logfc_threshold) %>%
    filter(!gene %in% background_down)
  
  res_no_background <- rbind(res_up, res_down)
  return(res_no_background)
}

remove_gbp_background <- function(res, res_gbp){
  res_gbp_up <- res_gbp %>%
    filter(log2FoldChange > logfc_threshold) %>%
    filter(padj < 0.05)
  
  res_gbp_down <- res_gbp %>%
    filter(log2FoldChange < -logfc_threshold) %>%
    filter(padj < 0.05)
  
  background_up <- c(res_gbp_up$gene)
  background_down <- c(res_gbp_down$gene)
  
  res_up <- res %>%
    filter(log2FoldChange > logfc_threshold) %>%
    filter(!gene %in% background_up)
  
  res_down <- res %>%
    filter(log2FoldChange < -logfc_threshold) %>%
    filter(!gene %in% background_down)
  
  res_no_background <- rbind(res_up, res_down)
  return(res_no_background)
}

res_basket_nigg <- remove_igg_background(res_basket, res_igg)
res_total_ngbp <- remove_gbp_background(res_total, res_gbp)
res_basketless_ngbp <- remove_gbp_background(res_basketless, res_gbp)

# write.xlsx(res_total_ngbp, file="outputs/xlsx/deseq2_results_no_specific_background.xlsx", sheetName="res_total_ngbp", row.names=FALSE)
# write.xlsx(res_basket_nigg, file="outputs/xlsx/deseq2_results_no_specific_background.xlsx", sheetName="res_basket_nigg", row.names=FALSE, append = T)
# write.xlsx(res_basketless_ngbp, file="outputs/xlsx/deseq2_results_no_specific_background.xlsx", sheetName="res_basketless_ngbp", row.names=FALSE, append = T)
```

### This report was run with a log2FC threshold of `r logfc_threshold`

## Genes enriched and depleted in NPCs  

```{r upset_plot, fig.height = 4, fig.width = 10}
# upset plot after removal (nsb - no specific background) ----------------------
up_nsb <- fromList(list("total" = filter(res_total_ngbp, log2FoldChange > logfc_threshold)$gene,
                        'basket' = filter(res_basket_nigg, log2FoldChange > logfc_threshold)$gene,
                        'basketless' = filter(res_basketless_ngbp, log2FoldChange > logfc_threshold)$gene))

upset_up <- upset(up_nsb, nsets = 3, set_size.show = T, set_size.scale_max = 1800)
upset_up_cp<- cowplot::plot_grid(NULL, 
                                 upset_up$Main_bar, upset_up$Sizes, upset_up$Matrix,
                                 nrow=2, align='hv', rel_heights = c(3,1),
                                 rel_widths = c(2,3))

down_nsb <-  fromList(list("total" = filter(res_total_ngbp, log2FoldChange < -logfc_threshold)$gene,
                           'basket' = filter(res_basket_nigg, log2FoldChange < -logfc_threshold)$gene,
                           'basketless' = filter(res_basketless_ngbp, log2FoldChange < -logfc_threshold)$gene))
upset_down <- upset(down_nsb, nsets = 3, set_size.show = T, set_size.scale_max = 1800)
upset_down_cp <- cowplot::plot_grid(NULL, 
                                    upset_down$Main_bar, upset_down$Sizes, upset_down$Matrix,
                                    nrow=2, align='v', rel_heights = c(3,1),
                                    rel_widths = c(2, 3))

gridExtra::grid.arrange(upset_up_cp, upset_down_cp, ncol = 2,
                        top = grid::textGrob("Enriched                                                  Depleted"))
```

```{r separate_shared_sets}
# separate genes from 7 permutations, no specific background -------------------
fromList2 <- function (input) {
  # Same as original fromList()...
  elements <- unique(unlist(input))
  data <- unlist(lapply(input, function(x) {
    x <- as.vector(match(elements, x))
  }))
  data[is.na(data)] <- as.integer(0)
  data[data != 0] <- as.integer(1)
  data <- data.frame(matrix(data, ncol = length(input), byrow = F))
  data <- data[which(rowSums(data) != 0), ]
  names(data) <- names(input)
  # ... Except now it conserves your original value names!
  row.names(data) <- elements
  return(data)
}

get_intersect_members <- function (x, ...){
  require(dplyr)
  require(tibble)
  x <- x[,sapply(x, is.numeric)][,0<=colMeans(x[,sapply(x, is.numeric)],na.rm=T) & colMeans(x[,sapply(x, is.numeric)],na.rm=T)<=1]
  n <- names(x)
  x %>% rownames_to_column() -> x
  l <- c(...)
  a <- intersect(names(x), l)
  ar <- vector('list',length(n)+1)
  ar[[1]] <- x
  i=2
  for (item in n) {
    if (item %in% a){
      if (class(x[[item]])=='integer'){
        ar[[i]] <- paste(item, '>= 1')
        i <- i + 1
      }
    } else {
      if (class(x[[item]])=='integer'){
        ar[[i]] <- paste(item, '== 0')
        i <- i + 1
      }
    }
  }
  do.call(filter_, ar) %>% column_to_rownames() -> x
  return(x)
}

format_intersect_members <- function(intersect_members_df){
  intersect_members_df <- intersect_members_df %>%
    rownames_to_column("gene") %>%
    mutate(common = orf_to_common(gene)) %>%
    mutate(description = orf_to_description(gene))
  return(intersect_members_df)
}

# res total minus background
# res basket minus background
# res basketless minus background
# res basket, basketless, total minus background
# res basket, total, minus background
# res basketless, total, minus background
# res basket, basketless, minus background

up_nsb2 <- fromList2(list("total" = filter(res_total_ngbp, log2FoldChange > logfc_threshold)$gene,
                          'basket' = filter(res_basket_nigg, log2FoldChange > logfc_threshold)$gene,
                          'basketless' = filter(res_basketless_ngbp, log2FoldChange > logfc_threshold)$gene))

total_basket_nsb_up <- format_intersect_members(get_intersect_members(up_nsb2, "total", "basket"))
total_basketless_nsb_up <- format_intersect_members(get_intersect_members(up_nsb2, "total", "basketless"))
total_basketless_basket_nsb_up <- format_intersect_members(get_intersect_members(up_nsb2, "total", "basket", "basketless"))
basket_nsb_up <- format_intersect_members(get_intersect_members(up_nsb2, "basket"))
basketless_nsb_up <- format_intersect_members(get_intersect_members(up_nsb2, "basketless"))
total_nsb_up <- format_intersect_members(get_intersect_members(up_nsb2, "total"))
basket_basketless_nsb_up <- format_intersect_members(get_intersect_members(up_nsb2, "basket", "basketless"))

# define set sizes for auto figure labelling by log2fc threshold
total_basket_nsb_up_size <- nrow(total_basket_nsb_up)
total_basketless_nsb_up_size <- nrow(total_basketless_nsb_up) 
total_basketless_basket_nsb_up_size <- nrow(total_basketless_basket_nsb_up)
basket_nsb_up_size <- nrow(basket_nsb_up)
basketless_nsb_up_size <- nrow(basketless_nsb_up)
total_nsb_up_size <- nrow(total_nsb_up)
basket_basketless_nsb_up_size <- nrow(basket_basketless_nsb_up)

# write.xlsx(total_basket_nsb_up, file="outputs/xlsx/upset_permutations_up_nsb.xlsx", sheetName="total_basket_nsb_up", row.names=FALSE)
# write.xlsx(total_basketless_nsb_up, file="outputs/xlsx/upset_permutations_up_nsb.xlsx", sheetName="total_basketless_nsb_up", append = T, row.names=FALSE)
# write.xlsx(total_basketless_basket_nsb_up, file="outputs/xlsx/upset_permutations_up_nsb.xlsx", sheetName="total_basketless_basket_nsb_up", append = T, row.names=FALSE)
# write.xlsx(basket_nsb_up, file="outputs/xlsx/upset_permutations_up_nsb.xlsx", sheetName="basket_nsb_up", append = T, row.names=FALSE)
# write.xlsx(basketless_nsb_up, file="outputs/xlsx/upset_permutations_up_nsb.xlsx", sheetName="basketless_nsb_up", append = T, row.names=FALSE)
# write.xlsx(total_nsb_up, file="outputs/xlsx/upset_permutations_up_nsb.xlsx", sheetName="total_nsb_up", append = T, row.names=FALSE)
# write.xlsx(basket_basketless_nsb_up, file="outputs/xlsx/upset_permutations_up_nsb.xlsx", sheetName="basket_basketless_nsb_up", append = T, row.names=FALSE)

down_nsb2 <- fromList2(list("total" = filter(res_total_ngbp, log2FoldChange < -logfc_threshold)$gene,
                            'basket' = filter(res_basket_nigg, log2FoldChange < -logfc_threshold)$gene,
                            'basketless' = filter(res_basketless_ngbp, log2FoldChange < -logfc_threshold)$gene))

total_basket_nsb_down <- format_intersect_members(get_intersect_members(down_nsb2, "total", "basket"))
total_basketless_nsb_down <- format_intersect_members(get_intersect_members(down_nsb2, "total", "basketless"))
total_basketless_basket_nsb_down <- format_intersect_members(get_intersect_members(down_nsb2, "total", "basket", "basketless"))
basket_nsb_down <- format_intersect_members(get_intersect_members(down_nsb2, "basket"))
basketless_nsb_down <- format_intersect_members(get_intersect_members(down_nsb2, "basketless"))
total_nsb_down <- format_intersect_members(get_intersect_members(down_nsb2, "total"))
basket_basketless_nsb_down <- format_intersect_members(get_intersect_members(down_nsb2, "basket", "basketless"))

# define set sizes for auto figure labelling by log2fc threshold
total_basket_nsb_down_size <- nrow(total_basket_nsb_down)
total_basketless_nsb_down_size <- nrow(total_basketless_nsb_down) 
total_basketless_basket_nsb_down_size <- nrow(total_basketless_basket_nsb_down)
basket_nsb_down_size <- nrow(basket_nsb_down)
basketless_nsb_down_size <- nrow(basketless_nsb_down)
total_nsb_down_size <- nrow(total_nsb_down)
basket_basketless_nsb_down_size <- nrow(basket_basketless_nsb_down)

# write.xlsx(total_basket_nsb_down, file="outputs/xlsx/upset_permutations_down_nsb.xlsx", sheetName="total_basket_nsb_down", row.names=FALSE)
# write.xlsx(total_basketless_nsb_down, file="outputs/xlsx/upset_permutations_down_nsb.xlsx", sheetName="total_basketless_nsb_down", append = T, row.names=FALSE)
# write.xlsx(total_basketless_basket_nsb_down, file="outputs/xlsx/upset_permutations_down_nsb.xlsx", sheetName="total_basketless_basket_nsb_down", append = T, row.names=FALSE)
# write.xlsx(basket_nsb_down, file="outputs/xlsx/upset_permutations_down_nsb.xlsx", sheetName="basket_nsb_down", append = T, row.names=FALSE)
# write.xlsx(basketless_nsb_down, file="outputs/xlsx/upset_permutations_down_nsb.xlsx", sheetName="basketless_nsb_down", append = T, row.names=FALSE)
# write.xlsx(total_nsb_down, file="outputs/xlsx/upset_permutations_down_nsb.xlsx", sheetName="total_nsb_down", append = T, row.names=FALSE)
# write.xlsx(basket_basketless_nsb_down, file="outputs/xlsx/upset_permutations_down_nsb.xlsx", sheetName="basket_basketless_nsb_down", append = T, row.names=FALSE)
```

**Question**: For things that are significantly associated with both basket and basketless, are they more likely to be associated with basket than basketless? 

**Answer**: Yes. For genes that are enriched, the log2FC values is significantly higher, and for genes that are depleted, the log2FC values are significantly lower (see t test estimate for size of log2FC difference).

```{r shared_set_genes_more_in_basket}
res_basket_nigg_shared <- res_basket_nigg %>%
  filter(gene %in% c(total_basketless_basket_nsb_up$gene, basket_basketless_nsb_up$gene)) %>%
  mutate(ap = "basket") 

res_basketless_ngbp_shared <- res_basketless_ngbp %>%
  filter(gene %in% c(total_basketless_basket_nsb_up$gene, basket_basketless_nsb_up$gene)) %>%
  mutate(ap = "basketless")

shared_up <- left_join(res_basket_nigg_shared, res_basketless_ngbp_shared, by = c("gene", "baseMean", "common", "description"))

plt1_log2fc <- ggplot(shared_up, aes(x = log2FoldChange.x, y = log2FoldChange.y, label = common)) +
  geom_point(alpha = .1) +
  geom_abline(slope = 1, intercept = 0) +
  theme_minimal() +
  theme(plot.title.position =  "plot",
        plot.title =element_text(size = 9)) +
  ylim(1, 5) +
  labs(x = "log2FC basket", y = "log2FC basketless", 
       title = paste0("Basket + Basketless, n = ", 
                      basket_basketless_nsb_up_size, " + ", 
                      total_basketless_basket_nsb_up_size))
                      

ttest1_log2fc <- t.test(x = shared_up$log2FoldChange.x, y = shared_up$log2FoldChange.y, paired = T) %>%
  tidy() %>%
  select(-statistic, -parameter, -alternative, -conf.low, -conf.high)

plt2_log2fc <- ggtexttable(ttest1_log2fc, rows = NULL, theme = ttheme(base_size = 7))

res_basket_nigg_shared <- res_basket_nigg %>%
  filter(gene %in% c(total_basketless_basket_nsb_down$gene, basket_basketless_nsb_down$gene)) %>%
  mutate(ap = "basket") 

res_basketless_ngbp_shared <- res_basketless_ngbp %>%
  filter(gene %in% c(total_basketless_basket_nsb_down$gene, basket_basketless_nsb_down$gene)) %>%
  mutate(ap = "basketless")

shared_down <- left_join(res_basket_nigg_shared, res_basketless_ngbp_shared, by = c("gene", "baseMean", "common", "description"))

plt3_log2fc <- ggplot(shared_down, aes(x = log2FoldChange.x, y = log2FoldChange.y, label = common)) +
  geom_point(alpha = .1) +
  geom_abline(slope = 1, intercept = 0) +
  theme_minimal() +
  theme(plot.title.position =  "plot",
        plot.title =element_text(size = 9)) +
  ylim(-6, -1) +
  labs(x = "log2FC basket", y = "log2FC basketless",
       title = paste0("Basket + Basketless, n = ", 
                      basket_basketless_nsb_down_size, " + ", 
                      total_basketless_basket_nsb_down_size))

ttest2_log2fc <- t.test(x = shared_down$log2FoldChange.x, y = shared_down$log2FoldChange.y, paired = T)%>%
  tidy() %>%
  select(-statistic, -parameter, -alternative, -conf.low, -conf.high)

plt4_log2fc <- ggtexttable(ttest2_log2fc, rows = NULL, theme = ttheme(base_size = 7))

ggarrange(plt1_log2fc, plt3_log2fc, plt2_log2fc,  plt4_log2fc, nrow = 2, ncol = 2)
```

[//]: # (FEATURES)

```{r define_features}
gtf <- import("GCF_000146045.2_R64_genomic.gtf")

features <- gtf %>%
  as.data.frame() %>%
  filter(type == "gene")

introns <- gtf %>% 
  as.data.frame() %>%
  filter(type == "exon") %>%
  group_by(gene_id) %>% 
  filter(n() > 1) %>% 
  ungroup() %>%
  dplyr::select(gene_id)
introns <- unique(introns$gene_id)

features <- features %>%
  mutate(introns = ifelse(gene_id %in% introns, "yes", "no"))

features <- features %>%
  dplyr::select(gene_id, width, strand, pseudo, introns)

norm_dds <- estimateSizeFactors(dds)
norm_counts <- DESeq2::counts(norm_dds, normalized=T)
polya_counts <- norm_counts %>%
  as.data.frame() %>%
  rownames_to_column("gene") %>%
  select(gene, starts_with("AP_G")) %>%
  rowwise() %>% 
  mutate(mean_polya = mean(c(AP_G_replicat1, AP_G_replicat2, AP_G_replicat3))) %>%
  select(gene, mean_polya)

vsd <- vst(dds)
vsd <- assay(vsd)
polya_vsd <- vsd %>%
  as.data.frame() %>%
  rownames_to_column("gene") %>%
  select(gene, starts_with("AP_G")) %>%
  rowwise() %>% 
  mutate(mean_polya_vst = mean(c(AP_G_replicat1, AP_G_replicat2, AP_G_replicat3))) %>%
  select(gene, mean_polya_vst)

features <- features %>%
  left_join(polya_counts, by = c("gene_id" = "gene")) %>%
  left_join(polya_vsd, by = c("gene_id" = "gene"))

# polyA
# 0.01 indicates transcripts that couldn't be measured bc there were fewer than 
# 5 transcripts in the library
polya_tails <- read_csv("https://data.mendeley.com/public-files/datasets/v5vm3dmm8y/files/aceb7ee8-5ba0-4cc5-be02-bdab3dc94cb3/file_downloaded") %>%
  select(Systematic.Name, polya_length = WT.BY.mean)

features <- left_join(features, polya_tails, by = c("gene_id" = "Systematic.Name"))

# half life
rates <- read_tsv("https://rnajournal.cshlp.org/content/suppl/2014/08/08/rna.045104.114.DC1/TableS5.xls") %>%
  select(gene_id = Syst, thalf, syn)
features <- left_join(features, rates, by = "gene_id")
```

```{r annotate_sets_with_feature}
res_igg_features <- res_igg %>%
  left_join(features, by = c("gene" = "gene_id")) %>%
  mutate(change = ifelse(log2FoldChange > 0 , "log2fc_pos", "log2fc_neg"))

res_gbp_features <- res_gbp %>%
  left_join(features, by = c("gene" = "gene_id")) %>%
  mutate(change = ifelse(log2FoldChange > 0 , "log2fc_pos", "log2fc_neg"))

res_basket_features <- res_basket_nigg %>%
  left_join(features, by = c("gene" = "gene_id")) %>%
  mutate(change = ifelse(log2FoldChange > 0 , "log2fc_pos", "log2fc_neg"))

res_total_features <- res_total_ngbp %>%
  left_join(features, by = c("gene" = "gene_id")) %>%
  mutate(change = ifelse(log2FoldChange > 0 , "log2fc_pos", "log2fc_neg"))

res_basketless_features <- res_basketless_ngbp %>%
  left_join(features, by = c("gene" = "gene_id")) %>%
  mutate(change = ifelse(log2FoldChange > 0 , "log2fc_pos", "log2fc_neg"))
```

```{r make_res_combined}
res_combined_features <- bind_rows(res_basket_features, res_basketless_features, res_total_features) %>%
  select(gene, common, description, width, strand, pseudo, introns, mean_polya, 
         mean_polya_vst, polya_length, thalf, syn, change) %>%
  distinct()
```

## Length of transcripts

**Question**: Does transcript length impact transcript association with NPCs?

**Answer**: Yes. Longer transcripts are more likely to associate with basket and basketless pores. The opposite pattern was observed for background controls IgG and GFP, where shorter transcripts were more likely to be enriched in these APs. 

While longer transcripts are more likely to be associated with both basket and basketless NPCs, the trend is more pronounced for transcripts that are enriched in basket and basketless. 

Brainstorming thought -- This may suggest that longer transcripts are more liklye to associate with pores due to chance -- because they're longer, they may be physically more likely to make contact with any pore. For the basketless pores, without other biological forces impacting which transcripts associate, this may be what drives this trend? 

Note -- I did the stats for strand as well as a control. Genes from one strand are not more/less likely to be associated NPCs.

### full sets

```{r LENGTH_full_sets}
# wilcox.test() -- wilcoxon rank sum test. Is mean gene length different
#                  between two groups? (data are not normally distributed)

igg_wilcox_length <- wilcox.test(res_igg_features$width ~ res_igg_features$change) %>%
  broom::tidy() %>%
  mutate(set = "IGG")

gbp_wilcox_length <- wilcox.test(res_gbp_features$width ~ res_gbp_features$change) %>%
  broom::tidy() %>%
  mutate(set = "GBP")

basket_wilcox_length <- wilcox.test(res_basket_features$width ~ res_basket_features$change) %>%
  broom::tidy() %>%
  mutate(set = "basket nsb")

basketless_wilcox_length <- wilcox.test(res_basketless_features$width ~ res_basketless_features$change) %>%
  broom::tidy() %>%
  mutate(set = "basketless nsb")

total_wilcox_length <- wilcox.test(res_total_features$width ~ res_total_features$change) %>%
  broom::tidy() %>%
  mutate(set = "total nsb")

all_wilcox_length <- rbind(igg_wilcox_length, 
                           gbp_wilcox_length,
                           basket_wilcox_length,
                           basketless_wilcox_length, 
                           total_wilcox_length) %>%
  as.data.frame() %>%
  mutate(method = "wilcox.test()") %>%
  dplyr::select(statistic, p.value, method, set)
  
plt1_length <- ggplot(res_basketless_features, aes(x = width)) +
  geom_density(alpha = .5, aes(fill = change)) +
  geom_density(data = features, aes(x = width), linetype = "dashed", color ="grey") +
  theme_minimal() +
  ggtitle("Basketless - GFP") + 
  xlim(0, 15000) +
  ylim(0, 0.0011)

plt2_length <- ggplot(res_total_features, aes(x = width)) +
  geom_density(alpha = .5, aes(fill = change)) +
  geom_density(data = features, aes(x = width), linetype = "dashed", color ="grey") +
  theme_minimal() +
  ggtitle("Total - GFP")+ 
  xlim(0, 15000) +
  ylim(0, 0.0011)

plt3_length <- ggplot(res_basket_features, aes(x = width)) +
  geom_density(alpha = .5, aes(fill = change)) +
  geom_density(data = features, aes(x = width), linetype = "dashed", color ="grey") +
  theme_minimal() +
  ggtitle("Basket - IgG")+ 
  xlim(0, 15000) +
  ylim(0, 0.0011)

plt4_length <- ggplot(res_igg_features, aes(x = width)) +
  geom_density(alpha = .5, aes(fill = change)) +
  geom_density(data = features, aes(x = width), linetype = "dashed", color ="grey") +
  theme_minimal() +
  ggtitle("IgG")+ 
  xlim(0, 15000) +
  ylim(0, 0.0011)

plt5_length <- ggplot(res_gbp_features, aes(x = width)) +
  geom_density(alpha = .5, aes(fill = change)) +
  geom_density(data = features, aes(x = width), linetype = "dashed", color ="grey") +
  theme_minimal() +
  ggtitle("GFP") + 
  xlim(0, 15000) +
  ylim(0, 0.0011)

plt6_length <- ggtexttable(all_wilcox_length, rows = NULL, theme = ttheme(base_size = 7))

plt7_length <- ggplot(features, aes(x = width)) +
  theme_minimal() +
  geom_density() +
  labs(title = "transcriptome distribution", x = "transcript length")

ggarrange1 <- ggarrange(plt1_length, plt2_length, plt3_length, 
                        nrow = 3, legend = "none", common.legend = T)
ggarrange2 <- ggarrange(plt4_length, plt5_length, plt6_length,
                        common.legend = T, nrow = 3, legend = "bottom")

ggarrange(ggarrange1, ggarrange2, ncol = 2, common.legend = T)
```

### shared 

```{r LENGTH_by_shared_set}
# wilcox.test() -- wilcoxon rank sum test. Is mean gene length different
#                  between two groups? (data are not normally distributed)

all_basket_basketless_wilcox_length2 <- res_basketless_features %>%
  filter(gene %in% c(total_basketless_basket_nsb_up$gene, 
                     total_basketless_basket_nsb_down$gene)) %>%
  do(tidy(wilcox.test(width ~ change, data = .))) %>%
  mutate(set = "basket, basketless, all")

basket_wilcox_length2 <- res_basket_features %>%
  filter(gene %in% c(basket_nsb_up$gene, basket_nsb_down$gene)) %>%
  do(tidy(wilcox.test(width ~ change, data = .))) %>%
  mutate(set = "basket")

basket_all_wilcox_length2 <- res_basket_features %>%
  filter(gene %in% c(total_basket_nsb_up$gene, 
                     total_basket_nsb_down$gene)) %>%
  do(tidy(wilcox.test(width ~ change, data = .))) %>%
  mutate(set = "basket, all")

all_wilcox_length2 <- rbind(basket_wilcox_length2,
                           basket_all_wilcox_length2,
                           all_basket_basketless_wilcox_length2) %>%
  as.data.frame() %>%
  mutate(method = "wilcox.test()") %>%
  dplyr::select(statistic, p.value, method, set)

plt1_length <- ggplot(res_basketless_features %>%
                filter(gene %in% c(total_basketless_basket_nsb_up$gene, 
                                   total_basketless_basket_nsb_down$gene)), 
              aes(x = width)) +
  geom_density(alpha = .5, aes(fill = change)) +
  geom_density(data = features, aes(x = width), linetype = "dashed", color ="black") +
  theme_minimal() +
  theme(plot.title = element_text(size=9),
        plot.title.position = "plot") +
  ggtitle(paste0("Basket, Basketless, All (n = ",
          total_basketless_basket_nsb_up_size, " + ",
          total_basketless_basket_nsb_down_size, ")")) +
  xlim(0, 15000) +
  ylim(0, 0.0012)

plt2_length <- ggplot(res_basket_features %>%
                filter(gene %in% c(basket_nsb_up$gene, basket_nsb_down$gene)), 
              aes(x = width)) +
  geom_density(alpha = .5, aes(fill = change)) +
  geom_density(data = features, aes(x = width), linetype = "dashed", color ="black") +
  theme_minimal() +
  theme(plot.title = element_text(size=9),
        plot.title.position = "plot") +
  ggtitle(paste0("Basket (n = ",
                 basket_nsb_up_size, " + ", 
                 basket_nsb_down_size, ")")) +
  xlim(0, 15000) +
  ylim(0, 0.0012)

plt3_length <- ggplot(res_basket_features %>%
                filter(gene %in% c(total_basket_nsb_up$gene, 
                                   total_basket_nsb_down$gene)), 
              aes(x = width)) +
  geom_density(alpha = .5, aes(fill = change)) +
  geom_density(data = features, aes(x = width), linetype = "dashed", color ="black") +
  theme_minimal() +
  theme(plot.title = element_text(size=9),
        plot.title.position = "plot") +
  ggtitle(paste0("Basket, All (n = ", 
                 total_basket_nsb_up_size, " + ",
                 total_basket_nsb_down_size, ")")) +
  xlim(0, 15000) +
  ylim(0, 0.0012)

plt4_length <- ggplot(res_combined_features, aes(x = width)) +
  geom_density(alpha = .5, aes(fill = change)) +
  geom_density(data = features, aes(x = width), linetype = "dashed", color ="black") +
  theme_minimal() +
  theme(plot.title = element_text(size=9),
        plot.title.position = "plot") +
  ggtitle(paste0("Basket, Basketless, All (n = ",
                 basket_nsb_up_size, " + ",
                 total_nsb_up_size, " + ",
                 basketless_nsb_up_size, " + ",
                 total_basket_nsb_up_size, " + ",
                 total_basketless_nsb_up_size, " + ",
                 basket_basketless_nsb_up_size, " + ",
                 total_basketless_basket_nsb_up_size, " +\n",
                 basket_nsb_down_size, " + ",
                 total_nsb_down_size, " + ",
                 basketless_nsb_down_size, " + ",
                 total_basket_nsb_down_size, " + ",
                 total_basketless_nsb_down_size, " + ",
                 basket_basketless_nsb_down_size, " + ",
                 total_basketless_basket_nsb_down_size, ")")) +
  xlim(0, 15000) +
  ylim(0, 0.0012)

plt5_length <- ggplot(res_combined_features %>%
                        filter(!gene %in% c(basket_nsb_up$gene, basket_nsb_down$gene)),
              aes(x = width)) +
  geom_density(alpha = .5, aes(fill = change)) +
  geom_density(data = features, aes(x = width), linetype = "dashed", color ="black") +
  theme_minimal() +
  theme(plot.title = element_text(size=9),
        plot.title.position = "plot") +
  ggtitle(paste0("Everything except distinct basket (n = ",
                 total_nsb_up_size, " + ",
                 basketless_nsb_up_size, " + ",
                 total_basket_nsb_up_size, " + ",
                 total_basketless_nsb_up_size, " + ",
                 basket_basketless_nsb_up_size, " + ",
                 total_basketless_basket_nsb_up_size, " +\n",
                 total_nsb_down_size, " + ",
                 basketless_nsb_down_size, " + ",
                 total_basket_nsb_down_size, " + ",
                 total_basketless_nsb_down_size, " + ",
                 basket_basketless_nsb_down_size, " + ",
                 total_basketless_basket_nsb_down_size, ")")) +
  xlim(0, 15000) +
  ylim(0, 0.0012)

plt6_length <- ggplot(res_combined_features %>%
                        filter(gene %in% c(basket_nsb_up$gene, basket_nsb_down$gene,
                                           total_basket_nsb_up$gene, total_basket_nsb_down$gene)),
              aes(x = width)) +
  geom_density(alpha = .5, aes(fill = change)) +
  geom_density(data = features, aes(x = width), linetype = "dashed", color ="black") +
  theme_minimal() +
  theme(plot.title = element_text(size=9),
        plot.title.position = "plot") +
  ggtitle(paste0("Basket, All (n = ",
                 basket_nsb_up_size, " + ",
                 total_basket_nsb_up_size, " + ",
                 basket_nsb_down_size, " + ",
                 total_basket_nsb_down_size, ")")) +
  xlim(0, 15000) +
  ylim(0, 0.0012)

plt7_length <- ggtexttable(all_wilcox_length2, rows = NULL, theme = ttheme(base_size = 8))
  
# ggarrange(ggarrange(plt2_length, plt3_length, plt1_length,
#                     plt4_length, plt5_length, plt6_length,
#                     legend = "bottom", common.legend = T, nrow = 3, ncol = 2),
#           plt7_length, nrow = 2, heights = c(3, 1))

ggarrange(plt2_length, plt3_length, plt1_length,
          plt4_length, plt5_length, plt6_length,
          legend = "bottom", common.legend = T, nrow = 3, ncol = 2)
```

```{r LENGTH_by_shared_set2}
plt1_length2 <- ggplot(res_basket_features %>%
                filter(gene %in% c(basket_nsb_up$gene, 
                                   basket_nsb_down$gene)), 
              aes(x = width, y = log2FoldChange, color = introns)) +
  geom_point(alpha = .5) +
  theme_minimal() +
  theme(plot.title = element_text(size=9),
        plot.title.position = "plot") +
  ggtitle(paste0("Basket (n = ", 
                 basket_nsb_up_size, " + ",
                 basket_nsb_down_size, ")")) +
  labs(x = "length") +
  scale_color_brewer(palette = "Set1")

plt2_length2 <- ggplot(res_basketless_features %>%
                filter(gene %in% c(total_basketless_basket_nsb_up$gene, 
                                   total_basketless_basket_nsb_down$gene)), 
              aes(x = width, y = log2FoldChange, color = introns)) +
  geom_point(alpha = .5) +
  theme_minimal() +
  theme(plot.title = element_text(size=9),
        plot.title.position = "plot") +
  ggtitle(paste0("Basket, Basketless, All (n = ",
          total_basketless_basket_nsb_up_size, " + ",
          total_basketless_basket_nsb_down_size, ")")) +
  labs(x = "length") +
  scale_color_brewer(palette = "Set1")

ggarrange(plt1_length2, plt2_length2, common.legend = T, legend = "bottom")
```

```{r STRAND_full_sets, eval = F}
# chisq.test() -- chi squared test. Are genes with introns or from one strand
#                 more or less likely to be associated with NPCs?

## IGG
chisq.test(table(res_igg_features$change, as.character(res_igg_features$strand)))
# genes from a one strand are not more likely to be enriched in igg

## GBP
chisq.test(table(res_gbp_features$change, as.character(res_gbp_features$strand)))
# genes from a one strand are not more likely to be enriched in gbp

## BASKET with specific backgroun (IGG) subtracted 
chisq.test(table(res_basket_features$change, as.character(res_basket_features$strand)))
# genes from a one strand are not more likely to be enriched in basket

## BASKETLESS with specific background removed (GBP)
chisq.test(table(res_basketless_features$change, as.character(res_basketless_features$strand)))
# genes from a one strand are not more likely to be enriched in basketless

## TOTAL with specific background removed (GBP)
chisq.test(table(res_total_features$change, as.character(res_total_features$strand)))
# genes from a one strand are not more likely to be enriched in total
```

## Introns -- full sets

**Question**: Are transcripts with introns associated with NPCs?

**Answer**: Introns are significantly depleted across all NPCs, suggesting these transcripts are processed/exported/whatever by something else.

```{r INTRONS_full_sets}
# chisq.test() -- chi squared test. Are genes with introns 
#                 more or less likely to be associated with NPCs?

igg_intron_chisq <- chisq.test(table(res_igg_features$change, res_igg_features$introns)) %>%
  tidy() %>%
  mutate(set = "igg")

gbp_intron_chisq <- chisq.test(table(res_gbp_features$change, res_gbp_features$introns)) %>%
  tidy() %>%
  mutate(set = "gbp")

basket_intron_chisq <- chisq.test(table(res_basket_features$change, res_basket_features$introns)) %>%
  tidy() %>%
  mutate(set = "basket nsb")

basketless_intron_chisq <- chisq.test(table(res_basketless_features$change, res_basketless_features$introns)) %>%
  tidy() %>%
  mutate(set = "basketless nsb")

total_intron_chisq <- chisq.test(table(res_total_features$change, res_total_features$introns)) %>%
  tidy() %>%
  mutate(set = "total nsb")

all_intron_chisq <- rbind(igg_intron_chisq,
                          gbp_intron_chisq,
                          basketless_intron_chisq,
                          basket_intron_chisq,
                          total_intron_chisq) %>%
  as.data.frame() %>%
  dplyr::select(-parameter) %>%
  mutate(method = "chisq.test()")

plt1_introns <- ggplot(res_total_features) +
  aes(x = change, fill = introns) +
  geom_bar(stat = "count") +
  geom_text(aes(label = ..count..), stat = "count") +
  #scale_fill_hue() +
  theme_minimal() +
  ggtitle("Total - GFP") +
  scale_fill_brewer(palette = "Set1")

plt2_introns <- ggplot(res_basketless_features) +
  aes(x = change, fill = introns) +
  geom_bar(stat = "count") +
  geom_text(aes(label = ..count..), stat = "count") +
  theme_minimal() +
  ggtitle("Basketless - GFP")+
  scale_fill_brewer(palette = "Set1")

plt3_introns <- ggplot(res_basket_features) +
  aes(x = change, fill = introns) +
  geom_bar(stat = "count") +
  geom_text(aes(label = ..count..), stat = "count") +
  theme_minimal() +
  ggtitle("Basket - IgG") +
  scale_fill_brewer(palette = "Set1")

plt4_introns <- ggplot(res_igg_features) +
  aes(x = change, fill = introns) +
  geom_bar(stat = "count") +
  geom_text(aes(label = ..count..), stat = "count") +
  theme_minimal() +
  ggtitle("IgG") +
  scale_fill_brewer(palette = "Set1")

plt5_introns <- ggplot(res_gbp_features) +
  aes(x = change, fill = introns) +
  geom_bar(stat = "count") +
  geom_text(aes(label = ..count..), stat = "count") +
  theme_minimal() +
  ggtitle("GFP") +
  scale_fill_brewer(palette = "Set1")

plt6_introns <- ggtexttable(all_intron_chisq, rows = NULL, theme = ttheme(base_size = 8))

ggarrange1 <- ggarrange(plt1_introns, plt2_introns, plt3_introns, 
                        nrow = 3, legend = "none", common.legend = T)
ggarrange2 <- ggarrange(plt4_introns, plt5_introns, plt6_introns, common.legend = T, nrow =3,
                        legend = "bottom")

ggarrange(ggarrange1, ggarrange2, ncol = 2, common.legend = T)
```

**Check** What is the length of the intron-containing genes?

```{r INTRON_lengths}
plt1_intron_len <- ggplot(res_basketless_features %>%
                            filter(log2FoldChange < 0 ), aes(x = width, fill = introns)) +
  geom_density(alpha = .5) +
  scale_fill_brewer(palette = "Set1") +
  #geom_density(data = features, aes(x = width), linetype = "dashed", color ="black") +
  theme_minimal() +
  theme(plot.title = element_text(size=9),
        plot.title.position = "plot") +
  labs(x = "length", title = "Basketless - NSB, - log2FC")

plt2_intron_len <- ggplot(res_basket_features %>%
                            filter(log2FoldChange > 0 ), aes(x = width, fill = introns)) +
  geom_density(alpha = .5) +
  scale_fill_brewer(palette = "Set1") +
  #geom_density(data = features, aes(x = width), linetype = "dashed", color ="black") +
  theme_minimal() +
  theme(plot.title = element_text(size=9),
        plot.title.position = "plot") +
  labs(x = "length", title = "Basket - NSB, + log2FC")

ggarrange(plt2_intron_len, plt1_intron_len, common.legend = T, legend = "bottom")
```

```{r INTRONS_LENGTH_regression_model}
res_basket_features$introns2 <- ifelse(res_basket_features$introns == "yes", 1, 0)
basket_width_intron <- lm(formula = log2FoldChange ~ width + introns2, data = res_basket_features)
basket_width_intron <- lm(formula = log2FoldChange ~ width + introns2 + polya_length + mean_polya_vst, data = res_basket_features %>% filter(polya_length > 1), na.action = "na.omit")
basket_width_intron %>% glance()
basket_width_intron %>% tidy()

# res_basket_features %>%
#   select(gene, log2FoldChange, width, introns2, polya_length, mean_polya) %>%
#   drop_na() %>%
#   filter(polya_length > 1) %>%
#   group_by(introns2) %>%
#   tally()
# 
# res_basket_features %>%
#   dim()

res_basketless_features$introns2 <- ifelse(res_basketless_features$introns == "yes", 1, 0)
basketless_width_intron <- lm(formula = log2FoldChange ~ width + introns2 + polya_length + mean_polya_vst, data = res_basketless_features %>% filter(polya_length > 1), na.action = "na.omit")
basketless_width_intron %>% glance()
basketless_width_intron %>% tidy()

res_total_features$introns2 <- ifelse(res_total_features$introns == "yes", 1, 0)
total_width_intron <- lm(formula = log2FoldChange ~ width + introns2 + polya_length + mean_polya_vst, data = res_total_features)
total_width_intron %>% glance()
total_width_intron %>% tidy()
```

## Expression

**Question**: Does total expression of a transcript impact association with NPCs?

**Answer**: Transcripts that are enriched in basketless APs are more likely to be highly expressed. This suggests that these transcripts may associate with basketless pores due to chance -- transcripts that are highly expressed are more likely to associate to basketless pores by virtue of there being more copies of that transcript present in the cell at any given time. 

The opposite is true for transcripts that are only enriched in basket

### shared sets

For these tests, we defined two sets -- basket consists of transcripts that are only significantly associated with basket (n = `r basket_nsb_up_size + basket_nsb_down_size`) and all transcripts that are enriched in basketless (n = `r nrow(res_basketless_features)`). 

Additionally, it's tricky to work with total expression level of a transcript. 
Using expression in the poly(A) control library, we used variance stabilized transformation (VST) to normalize counts. This is what **vst** refers to in the plot legends.

Lastly, as can be seen above and from the scatter plots below, transcripts with introns are doing their own thing, so we filtered those out before performing any statistics because these transcripts are highly expressed and would likely skew comparisons. 

```{r EXPRESSION_by_shared_set1}
# it looks like transcripts that are more likely to be in basketless are 
# intron-less and highly expressed
# The opposite is true for transcripts that are only significantly enriched in basket 

plt1_exp_scatter <- ggplot(res_basketless_features, aes(x = log2FoldChange, y = mean_polya_vst, color = introns)) +
  geom_point(alpha = 0.25) +
  labs(title = "Basketless", y = "mean poly(A) VST counts") +
  theme_minimal() +
  scale_color_brewer(palette = "Set1")

plt1_exp_scatter_tbl <- res_basketless_features %>%
  group_by(introns, change) %>% 
  summarise(mean_polya_vst = round(mean(mean_polya_vst), digits = 2),
            mean_polya_counts = round(mean(mean_polya), digits = 2)) %>%
   ggtexttable(rows = NULL, theme = ttheme(base_size = 8))

plt2_exp_scatter <-  ggplot(res_basket_features %>%
         filter(gene %in% c(basket_nsb_up$gene, basket_nsb_down$gene)), 
       aes(x = log2FoldChange, y = mean_polya_vst, color = introns)) +
  labs(title = "Basket", y = "mean poly(A) VST counts") +
  geom_point(alpha = 0.25) +
  theme_minimal() +
  scale_color_brewer(palette = "Set1")

plt2_exp_scatter_tbl <- res_basket_features %>%
  filter(gene %in% c(basket_nsb_up$gene, basket_nsb_down$gene)) %>%
  group_by(introns, change) %>% 
  summarise(mean_polya_vst = round(mean(mean_polya_vst), digits = 2),
            mean_polya_counts = round(mean(mean_polya), digits = 2)) %>%
   ggtexttable(rows = NULL, theme = ttheme(base_size = 8))

plt3_exp_scatter <-  ggplot(res_basket_features %>%
         filter(gene %in% c(basket_nsb_up$gene, basket_nsb_down$gene)), 
       aes(x = log2FoldChange, y = mean_polya_vst, color = introns)) +
  labs(title = "Basket", y = "mean poly(A) VST counts") +
  geom_point(alpha = 0.25) +
  theme_minimal() +
  scale_color_brewer(palette = "Set1")

plt3_exp_scatter_tbl <- res_basket_features %>%
  filter(gene %in% c(basket_nsb_up$gene, basket_nsb_down$gene)) %>%
  group_by(introns, change) %>% 
  summarise(mean_polya_vst = round(mean(mean_polya_vst), digits = 2),
            mean_polya_counts = round(mean(mean_polya), digits = 2)) %>%
   ggtexttable(rows = NULL, theme = ttheme(base_size = 8))


ggarrange(plt2_exp_scatter, plt1_exp_scatter, 
          plt2_exp_scatter_tbl,  plt1_exp_scatter_tbl,
          common.legend = T, nrow = 2, ncol = 2, legend = "bottom", heights = c(2, 1))

```

```{r EXPRESSION_by_shared_set2}
# Stats
plt1_exp <- ggplot(res_basketless_features %>%
         filter(introns == "no"), 
       aes(x = mean_polya_vst, fill = change)) +
  geom_density(alpha = .5) +
  theme_minimal()+
  theme_minimal() +
  theme(plot.title = element_text(size = 9),
        plot.title.position = "plot") +
  labs(title = paste0("Basketless (n = ",
                      total_basketless_nsb_up_size, " + ",
                      basketless_nsb_up_size, " + ",
                      basket_basketless_nsb_up_size, " + ",
                      total_basketless_basket_nsb_up_size, " +\n",
                      total_basketless_nsb_down_size, " + ",
                      basketless_nsb_down_size, " + ",
                      basket_basketless_nsb_down_size, " + ",
                      total_basketless_basket_nsb_down_size, ")"),
       x = "mean poly(A) VST count")

plt3_exp <- res_basketless_features %>% 
  group_by(introns) %>% 
  do(tidy(t.test(mean_polya_vst ~ change, data = .))) %>%
  select(-estimate1, -estimate2, -parameter, -conf.low, -conf.high, -alternative) %>%
  ggtexttable(rows = NULL, theme = ttheme(base_size = 8))

## Look at distinctly enriched in basket
plt2_exp <- ggplot(res_basket_features %>%
         filter(gene %in% c(basket_nsb_up$gene, basket_nsb_down$gene)) %>%
         filter(introns == "no"), 
       aes(x = mean_polya_vst, fill = change)) +
  geom_density(alpha = 0.5) +
  theme_minimal() +
  theme(plot.title = element_text(size = 9),
        plot.title.position = "plot") +
  labs(title = paste0("Basket (n = ",
                      basket_nsb_up_size, " + ",
                      basket_nsb_down_size, ")"),
       x = "mean poly(A) VST count")

plt4_exp <- res_basket_features %>%
  filter(gene %in% c(basket_nsb_up$gene, basket_nsb_down$gene)) %>%
  group_by(introns) %>% 
  do(tidy(t.test(mean_polya_vst ~ change, data = .)))  %>%
  select(-estimate1, -estimate2, -parameter, -conf.low, -conf.high, -alternative) %>%
  ggtexttable(rows = NULL, theme = ttheme(base_size = 8))

ggarrange(plt2_exp, plt1_exp, plt4_exp, plt3_exp,
          common.legend = T, nrow = 2, ncol = 2, legend = "bottom", heights = c(2, 1))
```

## Poly(A) tail length

```{r POLYA_by_shared_set}
# wilcox.test() -- wilcoxon rank sum test. Is mean gene length different
#                  between two groups? (data are not normally distributed)

all_basket_basketless_wilcox_polya2 <- res_basketless_features %>%
  filter(gene %in% c(total_basketless_basket_nsb_up$gene, 
                     total_basketless_basket_nsb_down$gene)) %>%
  filter(polya_length > 1) %>%
  do(tidy(wilcox.test(polya_length ~ change, data = .))) %>%
  mutate(set = "basket, basketless, all")

basket_wilcox_polya2 <- res_basket_features %>%
  filter(gene %in% c(basket_nsb_up$gene, basket_nsb_down$gene)) %>%
  filter(polya_length > 1) %>%
  do(tidy(wilcox.test(polya_length ~ change, data = .))) %>%
  mutate(set = "basket")

basket_all_wilcox_polya2 <- res_basket_features %>%
  filter(gene %in% c(total_basket_nsb_up$gene, 
                     total_basket_nsb_down$gene)) %>%
  filter(polya_length > 1) %>%
  do(tidy(wilcox.test(polya_length ~ change, data = .))) %>%
  mutate(set = "basket, all")

all_wilcox_polya2 <- rbind(basket_wilcox_polya2,
                           basket_all_wilcox_polya2,
                           all_basket_basketless_wilcox_polya2) %>%
  as.data.frame() %>%
  mutate(method = "wilcox.test()") %>%
  dplyr::select(statistic, p.value, method, set)

plt1_polya <- ggplot(res_basketless_features %>%
                filter(gene %in% c(total_basketless_basket_nsb_up$gene, 
                                   total_basketless_basket_nsb_down$gene)) %>%
                        filter(polya_length > 1), 
              aes(x = polya_length)) +
  geom_density(alpha = .5, aes(fill = change)) +
  geom_density(data = features %>% filter(polya_length > 1), aes(x = polya_length), linetype = "dashed", color = "black") +
  theme_minimal() +
  theme(plot.title = element_text(size=9),
        plot.title.position = "plot") +
  ggtitle(paste0("Basket, Basketless, All (n = ",
          total_basketless_basket_nsb_up_size, " + ",
          total_basketless_basket_nsb_down_size, ")"))

plt2_polya <- ggplot(res_basket_features %>%
                filter(gene %in% c(basket_nsb_up$gene, basket_nsb_down$gene))%>%
                        filter(polya_length > 1), 
              aes(x = polya_length)) +
  geom_density(alpha = .5, aes(fill = change)) +
  geom_density(data = features %>% filter(polya_length > 1), aes(x = polya_length), linetype = "dashed", color = "black") +
  theme_minimal() +
  theme(plot.title = element_text(size=9),
        plot.title.position = "plot") +
  ggtitle(paste0("Basket (n = ",
                 basket_nsb_up_size, " + ", 
                 basket_nsb_down_size, ")")) 

plt3_polya <- ggplot(res_basket_features %>%
                        filter(gene %in% c(total_basket_nsb_up$gene, 
                                           total_basket_nsb_down$gene)) %>%
                        filter(polya_length > 1), 
              aes(x = polya_length)) +
  geom_density(alpha = .5, aes(fill = change)) +
  geom_density(data = features %>% filter(polya_length > 1), aes(x = polya_length), linetype = "dashed", color = "black") +
  theme_minimal() +
  theme(plot.title = element_text(size=9),
        plot.title.position = "plot") +
  ggtitle(paste0("Basket, All (n = ", 
                 total_basket_nsb_up_size, " + ",
                 total_basket_nsb_down_size, ")"))

res_combined_features <- bind_rows(res_basket_features, res_basketless_features, res_total_features) %>%
  select(gene, common, description, width, strand, pseudo, introns, mean_polya, 
         mean_polya_vst, polya_length, thalf, syn, change) %>%
  distinct()

plt4_polya <- ggplot(res_combined_features %>% filter(polya_length > 1), aes(x = polya_length)) +
  geom_density(alpha = .5, aes(fill = change)) +
  geom_density(data = features %>% filter(polya_length > 1), aes(x = polya_length), linetype = "dashed", color = "black") +
  theme_minimal() +
  theme(plot.title = element_text(size=9),
        plot.title.position = "plot") +
  ggtitle(paste0("Basket, Basketless, All (n = ",
                 basket_nsb_up_size, " + ",
                 total_nsb_up_size, " + ",
                 basketless_nsb_up_size, " + ",
                 total_basket_nsb_up_size, " + ",
                 total_basketless_nsb_up_size, " + ",
                 basket_basketless_nsb_up_size, " + ",
                 total_basketless_basket_nsb_up_size, " +\n",
                 basket_nsb_down_size, " + ",
                 total_nsb_down_size, " + ",
                 basketless_nsb_down_size, " + ",
                 total_basket_nsb_down_size, " + ",
                 total_basketless_nsb_down_size, " + ",
                 basket_basketless_nsb_down_size, " + ",
                 total_basketless_basket_nsb_down_size, ")"))

plt5_polya <- ggplot(res_combined_features %>%
                        filter(!gene %in% c(basket_nsb_up$gene, basket_nsb_down$gene)) %>%
                        filter(polya_length > 1), 
              aes(x = polya_length)) +
  geom_density(alpha = .5, aes(fill = change)) +
  geom_density(data = features %>% filter(polya_length > 1), aes(x = polya_length), linetype = "dashed", color = "black") +
  theme_minimal() +
  theme(plot.title = element_text(size=9),
        plot.title.position = "plot") +
  ggtitle(paste0("Everything except distinct basket (n = ",
                 total_nsb_up_size, " + ",
                 basketless_nsb_up_size, " + ",
                 total_basket_nsb_up_size, " + ",
                 total_basketless_nsb_up_size, " + ",
                 basket_basketless_nsb_up_size, " + ",
                 total_basketless_basket_nsb_up_size, " +\n",
                 total_nsb_down_size, " + ",
                 basketless_nsb_down_size, " + ",
                 total_basket_nsb_down_size, " + ",
                 total_basketless_nsb_down_size, " + ",
                 basket_basketless_nsb_down_size, " + ",
                 total_basketless_basket_nsb_down_size, ")"))

plt6_polya <- ggplot(res_combined_features %>%
                        filter(gene %in% c(basket_nsb_up$gene, basket_nsb_down$gene,
                                           total_basket_nsb_up$gene, total_basket_nsb_down$gene)) %>%
                        filter(polya_length > 1), 
              aes(x = polya_length)) +
  geom_density(alpha = .5, aes(fill = change)) +
  geom_density(data = features %>% filter(polya_length > 1), aes(x = polya_length), linetype = "dashed", color = "black") +
  theme_minimal() +
  theme(plot.title = element_text(size=9),
        plot.title.position = "plot") +
  ggtitle(paste0("Basket, All (n = ",
                 basket_nsb_up_size, " + ",
                 total_basket_nsb_up_size, " + ",
                 basket_nsb_down_size, " + ",
                 total_basket_nsb_down_size, ")"))


plt7_polya <- ggtexttable(all_wilcox_polya2, rows = NULL, theme = ttheme(base_size = 8))

plt8_polya <- ggplot(features %>%
                       filter(polya_length > 1), 
                     aes(x = polya_length)) +
  theme_minimal() +
  geom_density() +
  labs(title = "transcriptome distribution", x = "poly(A) tail length")

ggarrange(ggarrange(plt2_polya, plt3_polya, plt1_polya, plt4_polya, plt5_polya, plt6_polya,
                    legend = "bottom", common.legend = T, nrow = 3, ncol = 2),
          plt7_polya, nrow = 2, heights = c(4, 1))
```

```{r POLYA_2}
plt5_polya <- ggplot(features %>%
                      filter(polya_length > 1), 
                    aes(y = width, x = polya_length)) +
  geom_point(alpha = .1) +
  theme_minimal() +
  labs(y = "transcript length", x = "poly(A) tail length", title = "whole transcriptome")


plt6_polya <- ggplot(res_basketless_features %>%
                filter(gene %in% c(total_basketless_basket_nsb_up$gene, 
                                   total_basketless_basket_nsb_down$gene)) %>%
                        filter(polya_length > 1), 
             aes(x = polya_length, y = width)) +
  geom_point(alpha = .5, aes(color = change)) +
  theme_minimal() +
  theme(plot.title = element_text(size=9),
        plot.title.position = "plot") +
  ggtitle(paste0("Basket, Basketless, All (n = ",
          total_basketless_basket_nsb_up_size, " + ",
          total_basketless_basket_nsb_down_size, ")")) +
  labs(y = "transcript length", x = "poly(A) tail length")

plt7_polya <- ggplot(res_basket_features %>%
                filter(gene %in% c(basket_nsb_up$gene, basket_nsb_down$gene))%>%
                        filter(polya_length > 1), 
              aes(x = polya_length, y = width)) +
  geom_point(alpha = .5, aes(color = change)) +
  theme_minimal() +
  theme(plot.title = element_text(size=9),
        plot.title.position = "plot") +
  ggtitle(paste0("Basket (n = ",
                 basket_nsb_up_size, " + ", 
                 basket_nsb_down_size, ")"))  +
  labs(y = "transcript length", x = "poly(A) tail length")

plt8_polya<- ggplot(res_basket_features %>%
                        filter(gene %in% c(total_basket_nsb_up$gene, 
                                           total_basket_nsb_down$gene)) %>%
                        filter(polya_length > 1), 
              aes(x = polya_length, y = width)) +
  geom_point(alpha = .5, aes(color = change)) +
  theme_minimal() +
  theme(plot.title = element_text(size=9),
        plot.title.position = "plot") +
  ggtitle(paste0("Basket, All (n = ", 
                 total_basket_nsb_up_size, " + ",
                 total_basket_nsb_down_size, ")")) +
  labs(y = "transcript length", x = "poly(A) tail length")

ggarrange(plt5_polya, plt6_polya, plt7_polya, plt8_polya,
          common.legend = T, legend = "bottom")
```

## Synthesis rate

Note because the x axis is log10 transformed, synthesis time = 0 are NOT on the density plots.
I performed the statistics on non-log10 transformed synthesis rates.

It appears to me like there is some evidence that transcripts with shorter synthesis rates are more likely to be associated with basket-containing NPCs. At least the stats and density plots suggest this a *little*, but the scatter plots aren't super convincing. 

```{r SYNTHESIS_by_shared_set}
# wilcox.test() -- wilcoxon rank sum test. Is mean gene length different
#                  between two groups? (data are not normally distributed)

all_basket_basketless_wilcox_synth2 <- res_basketless_features %>%
  filter(gene %in% c(total_basketless_basket_nsb_up$gene, 
                     total_basketless_basket_nsb_down$gene)) %>%
  do(tidy(wilcox.test(syn ~ change, data = .))) %>%
  mutate(set = "basket, basketless, all")

basket_wilcox_synth2 <- res_basket_features %>%
  filter(gene %in% c(basket_nsb_up$gene, basket_nsb_down$gene)) %>%
  do(tidy(wilcox.test(syn ~ change, data = .))) %>%
  mutate(set = "basket")

basket_all_wilcox_synth2 <- res_basket_features %>%
  filter(gene %in% c(total_basket_nsb_up$gene, 
                     total_basket_nsb_down$gene)) %>%
  do(tidy(wilcox.test(syn ~ change, data = .))) %>%
  mutate(set = "basket, all")

all_wilcox_synth2 <- rbind(basket_wilcox_synth2,
                           basket_all_wilcox_synth2,
                           all_basket_basketless_wilcox_synth2) %>%
  as.data.frame() %>%
  mutate(method = "wilcox.test()") %>%
  dplyr::select(statistic, p.value, method, set)

plt1_synth <- ggplot(res_basketless_features %>%
                filter(gene %in% c(total_basketless_basket_nsb_up$gene, 
                                   total_basketless_basket_nsb_down$gene)), 
              aes(x = syn)) +
  geom_density(alpha = .5, aes(fill = change)) +
  geom_density(data = features, aes(x = syn), linetype = "dashed", color = "grey") +
  theme_minimal() +
  theme(plot.title = element_text(size=9),
        plot.title.position = "plot") +
  ggtitle(paste0("Basket, Basketless, All (n = ",
          total_basketless_basket_nsb_up_size, " + ",
          total_basketless_basket_nsb_down_size, ")"))+
  labs(x = "log10 synthesis rate") +
  scale_x_log10()

plt2_synth <- ggplot(res_basket_features %>%
                filter(gene %in% c(basket_nsb_up$gene, basket_nsb_down$gene)), 
              aes(x = syn)) +
  geom_density(alpha = .5, aes(fill = change)) +
  geom_density(data = features, aes(x = syn), linetype = "dashed", color = "grey") +
  theme_minimal() +
  theme(plot.title = element_text(size=9),
        plot.title.position = "plot") +
  ggtitle(paste0("Basket (n = ",
                 basket_nsb_up_size, " + ", 
                 basket_nsb_down_size, ")")) +
  labs(x = "log10 synthesis rate") +
  scale_x_log10()

plt3_synth <- ggplot(res_basket_features %>%
                        filter(gene %in% c(total_basket_nsb_up$gene, 
                                           total_basket_nsb_down$gene)), 
              aes(x = syn)) +
  geom_density(alpha = .5, aes(fill = change)) +
  geom_density(data = features, aes(x = syn), linetype = "dashed", color = "grey") +
  theme_minimal() +
  theme(plot.title = element_text(size=9),
        plot.title.position = "plot") +
  ggtitle(paste0("Basket, All (n = ", 
                 total_basket_nsb_up_size, " + ",
                 total_basket_nsb_down_size, ")")) +
  labs(x = "log10 synthesis rate") +
  scale_x_log10()

plt4_synth <- ggtexttable(all_wilcox_synth2, rows = NULL, theme = ttheme(base_size = 8))

plt5_synth <- ggplot(features, aes(x = syn)) +
  theme_minimal() +
  geom_density() +
  labs(title = "transcriptome distribution", x = "log10 synthesis rate") +
  scale_x_log10()

ggarrange(ggarrange(plt2_synth, plt3_synth, plt1_synth, legend = "bottom", common.legend = T, nrow = 1),
          plt4_synth, nrow = 2)
```

```{r SYNTHESIS_2}
plt6_synth <- ggplot(res_basketless_features %>%
                filter(gene %in% c(total_basketless_basket_nsb_up$gene, 
                                   total_basketless_basket_nsb_down$gene)), 
             aes(x = syn, y = log2FoldChange)) +
  geom_point(alpha = .1, aes(color = change)) +
  theme_minimal() +
  theme(plot.title = element_text(size=9),
        plot.title.position = "plot") +
  ggtitle(paste0("Basket, Basketless, All (n = ",
          total_basketless_basket_nsb_up_size, " + ",
          total_basketless_basket_nsb_down_size, ")")) +
  labs(y = "log2 Fold Change", x = "log10 synthesis rate") +
  scale_x_log10()

plt7_synth <- ggplot(res_basket_features %>%
                filter(gene %in% c(basket_nsb_up$gene, basket_nsb_down$gene)), 
              aes(x = syn, y = log2FoldChange)) +
  geom_point(alpha = .1, aes(color = change)) +
  theme_minimal() +
  theme(plot.title = element_text(size=9),
        plot.title.position = "plot") +
  ggtitle(paste0("Basket (n = ",
                 basket_nsb_up_size, " + ", 
                 basket_nsb_down_size, ")")) +
  labs(y = "log2 Fold Change", x = "log10 synthesis rate") +
  scale_x_log10()

plt8_synth<- ggplot(res_basket_features %>%
                        filter(gene %in% c(total_basket_nsb_up$gene, 
                                           total_basket_nsb_down$gene)),
              aes(x = syn, y = log2FoldChange)) +
  geom_point(alpha = .1, aes(color = change)) +
  theme_minimal() +
  theme(plot.title = element_text(size=9),
        plot.title.position = "plot") +
  ggtitle(paste0("Basket, All (n = ", 
                 total_basket_nsb_up_size, " + ",
                 total_basket_nsb_down_size, ")")) +
  labs(y = "log2 Fold Change", x = "log10 synthesis rate") +
  scale_x_log10()

ggarrange(plt6_synth, plt7_synth, plt8_synth, nrow = 1,
          common.legend = T, legend = "bottom")
```

## Half life

Transcripts with shorter half life are more likely to be associated with NPCs. 

```{r HALFLIFE_by_shared_set}
# wilcox.test() -- wilcoxon rank sum test. Is mean gene length different
#                  between two groups? (data are not normally distributed)

all_basket_basketless_wilcox_thalf2 <- res_basketless_features %>%
  filter(gene %in% c(total_basketless_basket_nsb_up$gene, 
                     total_basketless_basket_nsb_down$gene)) %>%
  do(tidy(wilcox.test(thalf ~ change, data = .))) %>%
  mutate(set = "basket, basketless, all")

basket_wilcox_thalf2 <- res_basket_features %>%
  filter(gene %in% c(basket_nsb_up$gene, basket_nsb_down$gene)) %>%
  do(tidy(wilcox.test(thalf ~ change, data = .))) %>%
  mutate(set = "basket")

basket_all_wilcox_thalf2 <- res_basket_features %>%
  filter(gene %in% c(total_basket_nsb_up$gene, 
                     total_basket_nsb_down$gene)) %>%
  do(tidy(wilcox.test(thalf ~ change, data = .))) %>%
  mutate(set = "basket, all")

all_wilcox_thalf2 <- rbind(basket_wilcox_thalf2,
                           basket_all_wilcox_thalf2,
                           all_basket_basketless_wilcox_thalf2) %>%
  as.data.frame() %>%
  mutate(method = "wilcox.test()") %>%
  dplyr::select(statistic, p.value, method, set)

plt1_thalf <- ggplot(res_basketless_features %>%
                filter(gene %in% c(total_basketless_basket_nsb_up$gene, 
                                   total_basketless_basket_nsb_down$gene)), 
              aes(x = thalf)) +
  geom_density(alpha = .5, aes(fill = change)) +
  geom_density(data = features, aes(x = thalf), linetype = "dashed", color = "grey") +
  theme_minimal() +
  theme(plot.title = element_text(size=9),
        plot.title.position = "plot") +
  ggtitle(paste0("Basket, Basketless, All (n = ",
          total_basketless_basket_nsb_up_size, " + ",
          total_basketless_basket_nsb_down_size, ")")) +
  labs(x = "thalf") 

plt2_thalf <- ggplot(res_basket_features %>%
                filter(gene %in% c(basket_nsb_up$gene, basket_nsb_down$gene)), 
              aes(x = thalf)) +
  geom_density(alpha = .5, aes(fill = change)) +
  geom_density(data = features, aes(x = thalf), linetype = "dashed", color = "grey") +
  theme_minimal() +
  theme(plot.title = element_text(size=9),
        plot.title.position = "plot") +
  ggtitle(paste0("Basket (n = ",
                 basket_nsb_up_size, " + ", 
                 basket_nsb_down_size, ")")) +
  labs(x = "thalf")

plt3_thalf <- ggplot(res_basket_features %>%
                        filter(gene %in% c(total_basket_nsb_up$gene, 
                                           total_basket_nsb_down$gene)), 
              aes(x = thalf)) +
  geom_density(alpha = .5, aes(fill = change)) +
  geom_density(data = features, aes(x = thalf), linetype = "dashed", color = "grey") +
  theme_minimal() +
  theme(plot.title = element_text(size=9),
        plot.title.position = "plot") +
  ggtitle(paste0("Basket, All (n = ", 
                 total_basket_nsb_up_size, " + ",
                 total_basket_nsb_down_size, ")")) +
  labs(x = "thalf") 

plt4_thalf <- ggtexttable(all_wilcox_thalf2, rows = NULL, theme = ttheme(base_size = 8))

ggarrange(ggarrange(plt2_thalf, plt3_thalf, plt1_thalf, legend = "bottom", common.legend = T, nrow = 1),
          plt4_thalf, nrow = 2)
```

```{r HALFLIFE_2}
plt6_thalf <- ggplot(res_basketless_features %>%
                filter(gene %in% c(total_basketless_basket_nsb_up$gene, 
                                   total_basketless_basket_nsb_down$gene)), 
             aes(x = thalf, y = log2FoldChange)) +
  geom_point(alpha = .1, aes(color = change)) +
  theme_minimal() +
  theme(plot.title = element_text(size=9),
        plot.title.position = "plot") +
  ggtitle(paste0("Basket, Basketless, All (n = ",
          total_basketless_basket_nsb_up_size, " + ",
          total_basketless_basket_nsb_down_size, ")")) 

plt7_thalf <- ggplot(res_basket_features %>%
                filter(gene %in% c(basket_nsb_up$gene, basket_nsb_down$gene)), 
              aes(x = thalf, y = log2FoldChange)) +
  geom_point(alpha = .1, aes(color = change)) +
  theme_minimal() +
  theme(plot.title = element_text(size=9),
        plot.title.position = "plot") +
  ggtitle(paste0("Basket (n = ",
                 basket_nsb_up_size, " + ", 
                 basket_nsb_down_size, ")"))

plt8_thalf<- ggplot(res_basket_features %>%
                        filter(gene %in% c(total_basket_nsb_up$gene, 
                                           total_basket_nsb_down$gene)),
              aes(x = thalf, y = log2FoldChange)) +
  geom_point(alpha = .1, aes(color = change)) +
  theme_minimal() +
  theme(plot.title = element_text(size=9),
        plot.title.position = "plot") +
  ggtitle(paste0("Basket, All (n = ", 
                 total_basket_nsb_up_size, " + ",
                 total_basket_nsb_down_size, ")")) 

ggarrange(plt6_thalf, plt7_thalf, plt8_thalf, nrow = 1,
          common.legend = T, legend = "bottom")
```

## Do all of these features correlate?

E.g. are we pulling the same pattern over and over?
Tl:dr, I don't think so, although there are some correlations that are good to be aware of.

```{r CORRELATE, fig.height=8, fig.width = 12}
library(GGally)

cor_basketless_features <- res_basketless_features %>%
  filter(gene %in% c(total_basketless_basket_nsb_up$gene, 
                     total_basketless_basket_nsb_down$gene)) %>%
  select(change, transcript_length = width, polya_length, log2FoldChange,
         syn, thalf, mean_polya_vst) 

ggpairs(cor_basketless_features, columns = 2:7,
       ggplot2::aes(colour=change, alpha = .25), showStrips = T,
       title = paste0("Basket, Basketless, All (n = ",
                      total_basketless_basket_nsb_up_size, " + ",
                      total_basketless_basket_nsb_down_size, ")")) +
  theme_minimal()

cor_basket_features <- res_basket_features %>%
  filter(gene %in% c(basket_nsb_up$gene, basket_nsb_down$gene)) %>%
  select(change, transcript_length = width, polya_length, log2FoldChange,
         syn, thalf, mean_polya_vst) 

ggpairs(cor_basket_features, columns = 2:7,
       ggplot2::aes(colour=change, alpha = .25), showStrips = T,
       title = paste0("Basket (n = ",
                      basket_nsb_up_size, " + ", 
                      basket_nsb_down_size, ")")) +
  theme_minimal()

cor_total_basket_features <- res_basket_features %>%
  filter(gene %in% c(total_basket_nsb_up$gene, 
                     total_basket_nsb_down$gene)) %>%
  select(change, transcript_length = width, polya_length, log2FoldChange,
         syn, thalf, mean_polya_vst) 

ggpairs(cor_total_basket_features, columns = 2:7,
       ggplot2::aes(colour=change, alpha = .25), showStrips = T,
       title= paste0("Basket, All (n = ", 
                     total_basket_nsb_up_size, " + ",
                     total_basket_nsb_down_size, ")")) +
  theme_minimal()
```

## Enrichment

**Question**: What are the transcripts that associated with NPCs?

**Answer**: I still don't really know if there is a super clear signal. Leaving these messy enrichment plots here to be mulled over. 

```{r enrichment_functions}
res_enrichment <- function(res, title){
  # separate to induced and repressed
  up <- res[res$log2FoldChange > 2, ]
  down <- res[res$log2FoldChange < -2, ] 
  
  # KEGG enrichment
  # ekegg_up <- enrichKEGG(up$gene, "sce", pAdjustMethod = "bonferroni") 
  # plt_rows <- ekegg_up@result %>%
  #   filter(p.adjust < .05) %>%
  #   nrow()
  # plt1 <- dotplot(ekegg_up, showCategory = plt_rows, font.size = 5) + theme_minimal(base_size = 5)
  # 
  # ekegg_down <- enrichKEGG(down$gene, "sce", pAdjustMethod = "bonferroni")
  # plt_rows <- ekegg_down@result %>%
  #   filter(p.adjust < .05) %>%
  #   nrow()
  # plt2 <- dotplot(ekegg_down, showCategory = plt_rows, font.size = 5) + theme_minimal(base_size = 5)
  
  # GO enrichment
  ego_up <- enrichGO(up$common, OrgDb = org.Sc.sgd.db, keyType = "GENENAME", 
                     ont = "ALL", pAdjustMethod = "bonferroni")
  plt_rows <- ego_up@result %>%
    filter(p.adjust < .05) %>%
    nrow()
  plt3 <- dotplot(ego_up, showCategory = plt_rows, font.size = 5) + theme_minimal(base_size = 5)
  
  ego_down <- enrichGO(down$common, OrgDb = org.Sc.sgd.db, keyType = "GENENAME", 
                       ont = "ALL", pAdjustMethod = "bonferroni")
  plt_rows <- ego_down@result %>%
    filter(p.adjust < .05) %>%
    nrow()
  plt4 <- dotplot(ego_down, showCategory = plt_rows, font.size = 5) + theme_minimal(base_size = 5)
  
  # combine plots
  all_plts <- ggarrange(plt3, plt4, ncol = 2, nrow = 1, common.legend = F, 
                       legend = "bottom", labels = c("A: log2FC +", "B: log2FC -"),
                       heights = c(1, 2))
  annotate_figure(all_plts, top = text_grob(title))
}

enrichment_intersect <- function(up, down, title){
  # GO enrichment
  require(org.Sc.sgd.db)
  ego_up <- enrichGO(up$common, OrgDb = org.Sc.sgd.db, keyType = "GENENAME", 
                     ont = "ALL", pAdjustMethod = "bonferroni")
  plt_rows <- ego_up@result %>%
    filter(p.adjust < .05) %>%
    nrow()
  plt1 <- dotplot(ego_up, showCategory = plt_rows, font.size = 6, split="ONTOLOGY") + 
    facet_grid(ONTOLOGY~., scale="free") +
    theme_minimal(base_size = 6)
  
  ego_down <- enrichGO(down$common, OrgDb = org.Sc.sgd.db, keyType = "GENENAME", 
                       ont = "ALL", pAdjustMethod = "bonferroni")
  plt_rows <- ego_down@result %>%
    filter(p.adjust < .05) %>%
    nrow()
  plt2 <- dotplot(ego_down, showCategory = plt_rows, font.size = 6, split="ONTOLOGY") + 
    facet_grid(ONTOLOGY~., scale="free") +
    theme_minimal(base_size = 6)
  # combine plots
  all_plts<- ggarrange(plt1, plt2, ncol = 2, nrow = 1, common.legend = F, 
                       legend = "bottom", labels = c("A: log2FC +", "B: log2FC -"),
                       heights = c(1, 2))
  annotate_figure(all_plts, top = text_grob(title))
}

rna_classes <- read_csv("https://osf.io/gqnwh/download", skip = 1) %>%
  dplyr::select("I", "II", "III", "IV", "V", "VI", "VII", "VIII", "IX", "X") %>%
  pivot_longer(cols = I:X, names_to = "rna_class", values_to = "gene") %>%
  filter(!grepl("CUT", gene)) %>%
  filter(!grepl("SUT", gene)) 

enrichment_intersect_rna_classes <- function(up, down, title){
  enricher_up <- enricher(gene = up$common, TERM2GENE = rna_classes, maxGSSize = 703)
  plt_rows <- enricher_up@result %>%
    filter(p.adjust < .05) %>%
    nrow()
  plt1 <- dotplot(enricher_up, showCategory = plt_rows, font.size = 8) + 
    theme_minimal(base_size = 6)
  
  enricher_down <- enricher(gene = down$common, TERM2GENE = rna_classes, maxGSSize = 703)
  plt_rows <- enricher_down@result %>%
    filter(p.adjust < .05) %>%
    nrow()
  plt2 <- dotplot(enricher_down, showCategory = plt_rows, font.size = 8) + 
    theme_minimal(base_size = 6)
  # combine plots
  all_plts<- ggarrange(plt1, plt2, ncol = 2, nrow = 1, common.legend = F, 
                       legend = "bottom", labels = c("A: log2FC +", "B: log2FC -"),
                       heights = c(1, 2))
  annotate_figure(all_plts, top = text_grob(title))
}
```

### full sets

```{r GO_ENRICHMENT_full_sets, fig.height = 12, width = 10}
# enrichment_intersect(up = filter(res_basket_nigg, log2FoldChange > 0),
#                     down = filter(res_basket_nigg, log2FoldChange < 0),
#                     title = "Basket NSB")

enrichment_intersect(up = filter(res_basketless_ngbp, log2FoldChange > 0),
                     down = filter(res_basketless_ngbp, log2FoldChange < 0),
                     title = "Basketless NSB")
```

```{r RNA_CLASSES_full_sets, fig.height = 2.5}
# enrichment_intersect_rna_classes(up = filter(res_basket_nigg, log2FoldChange > 0),
#                                  down = filter(res_basket_nigg, log2FoldChange < 0),
#                                  title = "Basket NSB")

enrichment_intersect_rna_classes(up = filter(res_basketless_ngbp, log2FoldChange > 0),
                                 down = filter(res_basketless_ngbp, log2FoldChange < 0),
                                 title = "Basketless NSB")
```

### shared sets

```{r GO_ENRICHMENT_shared_sets, fig.height = 6, width = 12}
#enrichment_intersect(up = total_basket_nsb_up, down=total_basket_nsb_down, title = "Total, Basket")
# enrichment_intersect(up = total_basketless_basket_nsb_up, down=total_basketless_basket_nsb_down, title = "Total, Basket, Basketless")

enrichment_intersect(up = basket_nsb_up, down=basket_nsb_down, title = "Basket")
```

```{r RNA_CLASSES_shared_sets, fig.height = 2.5}
# enrichment_intersect_rna_classes(up = total_basket_nsb_up, down=total_basket_nsb_down, title = "Total, Basket")
# enrichment_intersect_rna_classes(up = total_basketless_basket_nsb_up, down=total_basketless_basket_nsb_down, title = "Total, Basket, Basketless")
enrichment_intersect_rna_classes(up = basket_nsb_up, down=basket_nsb_down, title = "Basket")
```