---
title: "Bensidoun et al. 2021 AP-RNA-seq publication figures"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = F, message = F, include = T, eval = T, warning = F,
                      cache = F)
```

```{r libraries}
library(dplyr)
library(readr)
library(tibble)
library(tidyr)
library(ggplot2)
library(DESeq2)
library(UpSetR)
library(xlsx)
library(rtracklayer)
library(ggpmisc)
library(ggpubr)
library(broom)
library(kableExtra)
library(clusterProfiler)
```

```{r helper_functions}
orf_to_common <- function(keys){
  commons <- AnnotationDbi::select(org.Sc.sgd.db::org.Sc.sgd.db,
                                   keys = keys,
                                   columns=c("COMMON"),
                                   keytype="ORF")
  commons <- commons[match(unique(commons$ORF), commons$ORF), ]
  commons <- commons[ , -2]
  commons$COMMON <- ifelse(is.na(commons$COMMON), commons$ORF, commons$COMMON)
  commons <- unique(commons[ , ])
  commons <- commons$COMMON
  return(commons)
}

orf_to_description <- function(keys){
  description <- AnnotationDbi::select(org.Sc.sgd.db::org.Sc.sgd.db,
                                       keys = keys,
                                       columns=c("DESCRIPTION"),
                                       keytype="ORF")
  description <- description[match(unique(description$ORF), description$ORF), ]
  description <- description$DESCRIPTION
  return(description)
}
```

```{r diffex}
logfc_threshold = 1

info <- read_csv("inputs/info.csv")
info$condition<- factor(info$condition, levels = c("input", "basket", "basketless", 
                                                   "total", "igg", "gbp"))
counts <- read_tsv("outputs/counts/raw_counts.tsv") %>%
  filter(!grepl("__", gene)) %>%
  select(!starts_with("AP_D")) %>%
  column_to_rownames("gene")

dds <- DESeqDataSetFromMatrix(counts,
                              colData = info,
                              design = ~ condition)
ds <- DESeq(dds, test="Wald")
#resultsNames(ds)
# log2 fold change (MLE): condition total vs input
# results are log2(total/input)
# log2FC values are expressed as that change in gene expression in total compared
# to that in input.

res_basket <- results(ds, contrast = c("condition", "basket", "input"), alpha = .05)
res_basket <- res_basket %>%
  as.data.frame() %>%
  rownames_to_column("gene") %>%
  filter(padj < 0.05) %>%
  filter(abs(log2FoldChange) > logfc_threshold) %>%
  mutate(common = orf_to_common(gene)) %>%
  mutate(description = orf_to_description(gene))

res_basketless <- results(ds, contrast = c("condition", "basketless", "input"), alpha = .05)
res_basketless <- res_basketless %>%
  as.data.frame() %>%
  rownames_to_column("gene") %>%
  filter(padj < 0.05) %>%
  filter(abs(log2FoldChange) > logfc_threshold) %>%
  mutate(common = orf_to_common(gene)) %>%
  mutate(description = orf_to_description(gene))

res_total <- results(ds, contrast = c("condition", "total", "input"), alpha = .05)
res_total <- res_total %>%
  as.data.frame() %>%
  rownames_to_column("gene") %>%
  filter(padj < 0.05) %>%
  filter(abs(log2FoldChange) > logfc_threshold)  %>%
  mutate(common = orf_to_common(gene)) %>%
  mutate(description = orf_to_description(gene))

res_gbp <- results(ds, contrast = c("condition", "gbp", "input"), alpha = .05)
res_gbp <- res_gbp %>%
  as.data.frame() %>%
  rownames_to_column("gene") %>%
  filter(padj < 0.05) %>%
  filter(abs(log2FoldChange) > logfc_threshold)  %>%
  mutate(common = orf_to_common(gene)) %>%
  mutate(description = orf_to_description(gene))

res_igg <- results(ds, contrast = c("condition", "igg", "input"), alpha = .05)
res_igg <- res_igg %>%
  as.data.frame() %>%
  rownames_to_column("gene") %>%
  filter(padj < 0.05) %>%
  filter(abs(log2FoldChange) > logfc_threshold)  %>%
  mutate(common = orf_to_common(gene)) %>%
  mutate(description = orf_to_description(gene))

# write.xlsx(res_total, file="outputs/xlsx/deseq2_results_background.xlsx", sheetName="res_total", row.names=FALSE)
# write.xlsx(res_basket, file="outputs/xlsx/deseq2_results_background.xlsx", sheetName="res_basket", row.names=FALSE, append = T)
# write.xlsx(res_basketless, file="outputs/xlsx/deseq2_results_background.xlsx", sheetName="res_basketless", row.names=FALSE, append = T)
# write.xlsx(res_igg, file="outputs/xlsx/deseq2_results_background.xlsx", sheetName="res_igg", row.names=FALSE, append = T)
# write.xlsx(res_gbp, file="outputs/xlsx/deseq2_results_background.xlsx", sheetName="res_gbp", row.names=FALSE, append = T)
```

```{r remove_specific_background}
# remove specific background from differentially expressed genes
remove_igg_background <- function(res, res_igg){
  res_igg_up <- res_igg %>%
    filter(log2FoldChange > logfc_threshold) %>%
    filter(padj < 0.05)
  
  res_igg_down <- res_igg %>%
    filter(log2FoldChange < -logfc_threshold) %>%
    filter(padj < 0.05)
  
  background_up <- c(res_igg_up$gene)
  background_down <- c(res_igg_down$gene)
  
  res_up <- res %>%
    filter(log2FoldChange > logfc_threshold) %>%
    filter(!gene %in% background_up)
  
  res_down <- res %>%
    filter(log2FoldChange < -logfc_threshold) %>%
    filter(!gene %in% background_down)
  
  res_no_background <- rbind(res_up, res_down)
  return(res_no_background)
}

remove_gbp_background <- function(res, res_gbp){
  res_gbp_up <- res_gbp %>%
    filter(log2FoldChange > logfc_threshold) %>%
    filter(padj < 0.05)
  
  res_gbp_down <- res_gbp %>%
    filter(log2FoldChange < -logfc_threshold) %>%
    filter(padj < 0.05)
  
  background_up <- c(res_gbp_up$gene)
  background_down <- c(res_gbp_down$gene)
  
  res_up <- res %>%
    filter(log2FoldChange > logfc_threshold) %>%
    filter(!gene %in% background_up)
  
  res_down <- res %>%
    filter(log2FoldChange < -logfc_threshold) %>%
    filter(!gene %in% background_down)
  
  res_no_background <- rbind(res_up, res_down)
  return(res_no_background)
}

res_basket_nigg <- remove_igg_background(res_basket, res_igg)
res_total_ngbp <- remove_gbp_background(res_total, res_gbp)
res_basketless_ngbp <- remove_gbp_background(res_basketless, res_gbp)

# write.xlsx(res_total_ngbp, file="outputs/xlsx/deseq2_results_no_specific_background.xlsx", sheetName="res_total_ngbp", row.names=FALSE)
# write.xlsx(res_basket_nigg, file="outputs/xlsx/deseq2_results_no_specific_background.xlsx", sheetName="res_basket_nigg", row.names=FALSE, append = T)
# write.xlsx(res_basketless_ngbp, file="outputs/xlsx/deseq2_results_no_specific_background.xlsx", sheetName="res_basketless_ngbp", row.names=FALSE, append = T)
```

### This report was run with a log2FC threshold of `r logfc_threshold`

## Upset plots (Figure 7A & 7B, left)

```{r upset_plot, fig.height = 4, fig.width = 10}
# upset plot after removal (nsb - no specific background) ----------------------
up_nsb <- fromList(list("all pore" = filter(res_total_ngbp, log2FoldChange > logfc_threshold)$gene,
                        'basket' = filter(res_basket_nigg, log2FoldChange > logfc_threshold)$gene,
                        'basketless' = filter(res_basketless_ngbp, log2FoldChange > logfc_threshold)$gene))

upset_up <- upset(up_nsb, nsets = 3, set_size.show = T, set_size.scale_max = 1800)
upset_up_cp<- cowplot::plot_grid(NULL, 
                                 upset_up$Main_bar, upset_up$Sizes, upset_up$Matrix,
                                 nrow=2, align='hv', rel_heights = c(3,1),
                                 rel_widths = c(2,3))

down_nsb <-  fromList(list("all pore" = filter(res_total_ngbp, log2FoldChange < -logfc_threshold)$gene,
                           'basket' = filter(res_basket_nigg, log2FoldChange < -logfc_threshold)$gene,
                           'basketless' = filter(res_basketless_ngbp, log2FoldChange < -logfc_threshold)$gene))
upset_down <- upset(down_nsb, nsets = 3, set_size.show = T, set_size.scale_max = 1800)
upset_down_cp <- cowplot::plot_grid(NULL, 
                                    upset_down$Main_bar, upset_down$Sizes, upset_down$Matrix,
                                    nrow=2, align='v', rel_heights = c(3,1),
                                    rel_widths = c(2, 3))

gridExtra::grid.arrange(upset_up_cp, upset_down_cp, ncol = 2,
                        top = grid::textGrob("Enriched                                                  Depleted"))
```

```{r save_upset_as_pdf, include = F, message=F}
pdf("publication_figures/upset_up.pdf", width =4.3, height = 3.5)
upset_up
dev.off()

png("publication_figures/upset_up.png", units = "in", res = 300, width =4.3, height = 3.5)
upset_up
dev.off()

pdf("publication_figures/upset_down.pdf", width = 4.3, height = 3.5)
upset_down
dev.off()

png("publication_figures/upset_down.png", units = "in", res = 300, width = 4.3, height = 3.5)
upset_down
dev.off()
```

```{r separate_shared_sets}
# separate genes from 7 permutations, no specific background -------------------
fromList2 <- function (input) {
  # Same as original fromList()...
  elements <- unique(unlist(input))
  data <- unlist(lapply(input, function(x) {
    x <- as.vector(match(elements, x))
  }))
  data[is.na(data)] <- as.integer(0)
  data[data != 0] <- as.integer(1)
  data <- data.frame(matrix(data, ncol = length(input), byrow = F))
  data <- data[which(rowSums(data) != 0), ]
  names(data) <- names(input)
  # ... Except now it conserves your original value names!
  row.names(data) <- elements
  return(data)
}

get_intersect_members <- function (x, ...){
  require(dplyr)
  require(tibble)
  x <- x[,sapply(x, is.numeric)][,0<=colMeans(x[,sapply(x, is.numeric)],na.rm=T) & colMeans(x[,sapply(x, is.numeric)],na.rm=T)<=1]
  n <- names(x)
  x %>% rownames_to_column() -> x
  l <- c(...)
  a <- intersect(names(x), l)
  ar <- vector('list',length(n)+1)
  ar[[1]] <- x
  i=2
  for (item in n) {
    if (item %in% a){
      if (class(x[[item]])=='integer'){
        ar[[i]] <- paste(item, '>= 1')
        i <- i + 1
      }
    } else {
      if (class(x[[item]])=='integer'){
        ar[[i]] <- paste(item, '== 0')
        i <- i + 1
      }
    }
  }
  do.call(filter_, ar) %>% column_to_rownames() -> x
  return(x)
}

format_intersect_members <- function(intersect_members_df){
  intersect_members_df <- intersect_members_df %>%
    rownames_to_column("gene") %>%
    mutate(common = orf_to_common(gene)) %>%
    mutate(description = orf_to_description(gene))
  return(intersect_members_df)
}

# res total minus background
# res basket minus background
# res basketless minus background
# res basket, basketless, total minus background
# res basket, total, minus background
# res basketless, total, minus background
# res basket, basketless, minus background

up_nsb2 <- fromList2(list("total" = filter(res_total_ngbp, log2FoldChange > logfc_threshold)$gene,
                          'basket' = filter(res_basket_nigg, log2FoldChange > logfc_threshold)$gene,
                          'basketless' = filter(res_basketless_ngbp, log2FoldChange > logfc_threshold)$gene))

total_basket_nsb_up <- format_intersect_members(get_intersect_members(up_nsb2, "total", "basket"))
total_basketless_nsb_up <- format_intersect_members(get_intersect_members(up_nsb2, "total", "basketless"))
total_basketless_basket_nsb_up <- format_intersect_members(get_intersect_members(up_nsb2, "total", "basket", "basketless"))
basket_nsb_up <- format_intersect_members(get_intersect_members(up_nsb2, "basket"))
basketless_nsb_up <- format_intersect_members(get_intersect_members(up_nsb2, "basketless"))
total_nsb_up <- format_intersect_members(get_intersect_members(up_nsb2, "total"))
basket_basketless_nsb_up <- format_intersect_members(get_intersect_members(up_nsb2, "basket", "basketless"))

# define set sizes for auto figure labelling by log2fc threshold
total_basket_nsb_up_size <- nrow(total_basket_nsb_up)
total_basketless_nsb_up_size <- nrow(total_basketless_nsb_up) 
total_basketless_basket_nsb_up_size <- nrow(total_basketless_basket_nsb_up)
basket_nsb_up_size <- nrow(basket_nsb_up)
basketless_nsb_up_size <- nrow(basketless_nsb_up)
total_nsb_up_size <- nrow(total_nsb_up)
basket_basketless_nsb_up_size <- nrow(basket_basketless_nsb_up)

# write.xlsx(total_basket_nsb_up, file="outputs/xlsx/upset_permutations_up_nsb.xlsx", sheetName="total_basket_nsb_up", row.names=FALSE)
# write.xlsx(total_basketless_nsb_up, file="outputs/xlsx/upset_permutations_up_nsb.xlsx", sheetName="total_basketless_nsb_up", append = T, row.names=FALSE)
# write.xlsx(total_basketless_basket_nsb_up, file="outputs/xlsx/upset_permutations_up_nsb.xlsx", sheetName="total_basketless_basket_nsb_up", append = T, row.names=FALSE)
# write.xlsx(basket_nsb_up, file="outputs/xlsx/upset_permutations_up_nsb.xlsx", sheetName="basket_nsb_up", append = T, row.names=FALSE)
# write.xlsx(basketless_nsb_up, file="outputs/xlsx/upset_permutations_up_nsb.xlsx", sheetName="basketless_nsb_up", append = T, row.names=FALSE)
# write.xlsx(total_nsb_up, file="outputs/xlsx/upset_permutations_up_nsb.xlsx", sheetName="total_nsb_up", append = T, row.names=FALSE)
# write.xlsx(basket_basketless_nsb_up, file="outputs/xlsx/upset_permutations_up_nsb.xlsx", sheetName="basket_basketless_nsb_up", append = T, row.names=FALSE)

down_nsb2 <- fromList2(list("total" = filter(res_total_ngbp, log2FoldChange < -logfc_threshold)$gene,
                            'basket' = filter(res_basket_nigg, log2FoldChange < -logfc_threshold)$gene,
                            'basketless' = filter(res_basketless_ngbp, log2FoldChange < -logfc_threshold)$gene))

total_basket_nsb_down <- format_intersect_members(get_intersect_members(down_nsb2, "total", "basket"))
total_basketless_nsb_down <- format_intersect_members(get_intersect_members(down_nsb2, "total", "basketless"))
total_basketless_basket_nsb_down <- format_intersect_members(get_intersect_members(down_nsb2, "total", "basket", "basketless"))
basket_nsb_down <- format_intersect_members(get_intersect_members(down_nsb2, "basket"))
basketless_nsb_down <- format_intersect_members(get_intersect_members(down_nsb2, "basketless"))
total_nsb_down <- format_intersect_members(get_intersect_members(down_nsb2, "total"))
basket_basketless_nsb_down <- format_intersect_members(get_intersect_members(down_nsb2, "basket", "basketless"))

# define set sizes for auto figure labelling by log2fc threshold
total_basket_nsb_down_size <- nrow(total_basket_nsb_down)
total_basketless_nsb_down_size <- nrow(total_basketless_nsb_down) 
total_basketless_basket_nsb_down_size <- nrow(total_basketless_basket_nsb_down)
basket_nsb_down_size <- nrow(basket_nsb_down)
basketless_nsb_down_size <- nrow(basketless_nsb_down)
total_nsb_down_size <- nrow(total_nsb_down)
basket_basketless_nsb_down_size <- nrow(basket_basketless_nsb_down)

# write.xlsx(total_basket_nsb_down, file="outputs/xlsx/upset_permutations_down_nsb.xlsx", sheetName="total_basket_nsb_down", row.names=FALSE)
# write.xlsx(total_basketless_nsb_down, file="outputs/xlsx/upset_permutations_down_nsb.xlsx", sheetName="total_basketless_nsb_down", append = T, row.names=FALSE)
# write.xlsx(total_basketless_basket_nsb_down, file="outputs/xlsx/upset_permutations_down_nsb.xlsx", sheetName="total_basketless_basket_nsb_down", append = T, row.names=FALSE)
# write.xlsx(basket_nsb_down, file="outputs/xlsx/upset_permutations_down_nsb.xlsx", sheetName="basket_nsb_down", append = T, row.names=FALSE)
# write.xlsx(basketless_nsb_down, file="outputs/xlsx/upset_permutations_down_nsb.xlsx", sheetName="basketless_nsb_down", append = T, row.names=FALSE)
# write.xlsx(total_nsb_down, file="outputs/xlsx/upset_permutations_down_nsb.xlsx", sheetName="total_nsb_down", append = T, row.names=FALSE)
# write.xlsx(basket_basketless_nsb_down, file="outputs/xlsx/upset_permutations_down_nsb.xlsx", sheetName="basket_basketless_nsb_down", append = T, row.names=FALSE)
```

## Stronger enrichment in basket (Figure 7A & 7B, right)

```{r shared_set_genes_more_in_basket, fig.height=3}
res_basket_nigg_shared <- res_basket_nigg %>%
  filter(gene %in% c(total_basketless_basket_nsb_up$gene, basket_basketless_nsb_up$gene)) %>%
  mutate(ap = "basket") 

res_basketless_ngbp_shared <- res_basketless_ngbp %>%
  filter(gene %in% c(total_basketless_basket_nsb_up$gene, basket_basketless_nsb_up$gene)) %>%
  mutate(ap = "basketless")

shared_up <- left_join(res_basket_nigg_shared, res_basketless_ngbp_shared, by = c("gene", "baseMean", "common", "description"))

plt1_log2fc <- ggplot(shared_up, aes(x = log2FoldChange.x, y = log2FoldChange.y, label = common)) +
  geom_point(alpha = .1) +
  geom_abline(slope = 1, intercept = 0) +
  theme_minimal() +
  theme(plot.title.position =  "plot",
        plot.title =element_text(size = 9)) +
  ylim(1, 5) +
  labs(x = expression(log[2]~fold~change~basket), y = expression(log[2]~fold~change~basketless), 
       title = paste0("Basket + Basketless, n = ", 
                      basket_basketless_nsb_up_size, " + ", 
                      total_basketless_basket_nsb_up_size))
                      
ttest1_log2fc <- t.test(x = shared_up$log2FoldChange.x, y = shared_up$log2FoldChange.y, paired = T) %>%
  tidy() %>%
  dplyr::select(-statistic, -parameter, -alternative, -conf.low, -conf.high)

plt2_log2fc <- ggtexttable(ttest1_log2fc, rows = NULL, theme = ttheme(base_size = 7))

res_basket_nigg_shared <- res_basket_nigg %>%
  filter(gene %in% c(total_basketless_basket_nsb_down$gene, basket_basketless_nsb_down$gene)) %>%
  mutate(ap = "basket") 

res_basketless_ngbp_shared <- res_basketless_ngbp %>%
  filter(gene %in% c(total_basketless_basket_nsb_down$gene, basket_basketless_nsb_down$gene)) %>%
  mutate(ap = "basketless")

shared_down <- left_join(res_basket_nigg_shared, res_basketless_ngbp_shared, by = c("gene", "baseMean", "common", "description"))

plt3_log2fc <- ggplot(shared_down, aes(x = log2FoldChange.x, y = log2FoldChange.y, label = common)) +
  geom_point(alpha = .1) +
  geom_abline(slope = 1, intercept = 0) +
  theme_minimal() +
  theme(plot.title.position =  "plot",
        plot.title =element_text(size = 9)) +
  ylim(-6, -1) +
  labs(x = expression(log[2]~fold~change~basket), y = expression(log[2]~fold~change~basketless),
       title = paste0("Basket + Basketless, n = ", 
                      basket_basketless_nsb_down_size, " + ", 
                      total_basketless_basket_nsb_down_size))

ttest2_log2fc <- t.test(x = shared_down$log2FoldChange.x, y = shared_down$log2FoldChange.y, paired = T)%>%
  tidy() %>%
  dplyr::select(-statistic, -parameter, -alternative, -conf.low, -conf.high)

plt4_log2fc <- ggtexttable(ttest2_log2fc, rows = NULL, theme = ttheme(base_size = 7))

ggarrange(plt1_log2fc, plt3_log2fc, nrow = 1, ncol = 2)
ggarrange(plt2_log2fc,  plt4_log2fc, nrow = 1, ncol = 2)
```

```{r save_shared_set_genes_more_in_basket_as_pdf, include = F, message=F}
pdf("publication_figures/shared_set_genes_more_in_basket_up.pdf", width = 2.5, height = 2.5)
plt1_log2fc
dev.off()

png("publication_figures/shared_set_genes_more_in_basket_up.png", units = "in", res = 300, width = 2.5, height = 2.5)
plt1_log2fc
dev.off()

pdf("publication_figures/shared_set_genes_more_in_basket_down.pdf", width = 2.5, height = 2.5)
plt3_log2fc
dev.off()

png("publication_figures/shared_set_genes_more_in_basket_down.png", units = "in", res = 300, width = 2.5, height = 2.5)
plt3_log2fc
dev.off()
```

<!-- ## Define Features -->

```{r define_features}
gtf <- import("GCF_000146045.2_R64_genomic.gtf")

features <- gtf %>%
  as.data.frame() %>%
  filter(type == "gene")

introns <- gtf %>% 
  as.data.frame() %>%
  filter(type == "exon") %>%
  group_by(gene_id) %>% 
  filter(n() > 1) %>% 
  ungroup() %>%
  dplyr::select(gene_id)
introns <- unique(introns$gene_id)

features <- features %>%
  mutate(introns = ifelse(gene_id %in% introns, "yes", "no"))

features <- features %>%
  dplyr::select(gene_id, width, strand, pseudo, introns)

norm_dds <- estimateSizeFactors(dds)
norm_counts <- DESeq2::counts(norm_dds, normalized=T)
polya_counts <- norm_counts %>%
  as.data.frame() %>%
  rownames_to_column("gene") %>%
  select(gene, starts_with("AP_G")) %>%
  rowwise() %>% 
  mutate(mean_polya = mean(c(AP_G_replicat1, AP_G_replicat2, AP_G_replicat3))) %>%
  select(gene, mean_polya)

vsd <- vst(dds)
vsd <- assay(vsd)
polya_vsd <- vsd %>%
  as.data.frame() %>%
  rownames_to_column("gene") %>%
  select(gene, starts_with("AP_G")) %>%
  rowwise() %>% 
  mutate(mean_polya_vst = mean(c(AP_G_replicat1, AP_G_replicat2, AP_G_replicat3))) %>%
  select(gene, mean_polya_vst)

features <- features %>%
  left_join(polya_counts, by = c("gene_id" = "gene")) %>%
  left_join(polya_vsd, by = c("gene_id" = "gene"))

# polyA
# 0.01 indicates transcripts that couldn't be measured bc there were fewer than 
# 5 transcripts in the library
polya_tails <- read_csv("https://data.mendeley.com/public-files/datasets/v5vm3dmm8y/files/aceb7ee8-5ba0-4cc5-be02-bdab3dc94cb3/file_downloaded") %>%
  select(Systematic.Name, polya_length = WT.BY.mean)

features <- left_join(features, polya_tails, by = c("gene_id" = "Systematic.Name"))

# half life
rates <- read_tsv("https://rnajournal.cshlp.org/content/suppl/2014/08/08/rna.045104.114.DC1/TableS5.xls") %>%
  select(gene_id = Syst, thalf, syn)
features <- left_join(features, rates, by = "gene_id")
```

```{r annotate_sets_with_feature}
res_igg_features <- res_igg %>%
  left_join(features, by = c("gene" = "gene_id")) %>%
  mutate(change = ifelse(log2FoldChange > 0 , "log2fc_pos", "log2fc_neg"))

res_gbp_features <- res_gbp %>%
  left_join(features, by = c("gene" = "gene_id")) %>%
  mutate(change = ifelse(log2FoldChange > 0 , "log2fc_pos", "log2fc_neg"))

res_basket_features <- res_basket_nigg %>%
  left_join(features, by = c("gene" = "gene_id")) %>%
  mutate(change = ifelse(log2FoldChange > 0 , "log2fc_pos", "log2fc_neg"))

res_total_features <- res_total_ngbp %>%
  left_join(features, by = c("gene" = "gene_id")) %>%
  mutate(change = ifelse(log2FoldChange > 0 , "log2fc_pos", "log2fc_neg"))

res_basketless_features <- res_basketless_ngbp %>%
  left_join(features, by = c("gene" = "gene_id")) %>%
  mutate(change = ifelse(log2FoldChange > 0 , "log2fc_pos", "log2fc_neg"))
```

```{r make_res_combined}
res_combined_features <- bind_rows(res_basket_features, res_basketless_features, res_total_features) %>%
  select(gene, common, description, width, strand, pseudo, introns, mean_polya, 
         mean_polya_vst, polya_length, thalf, syn, change) %>%
  distinct()
```

## VST Counts / poly(A) expression (Figure 7C)

```{r label_res_features_df_as_uniquely_enriched}
res_basketless_features <- res_basketless_features %>%
  mutate(distinct = ifelse(gene %in% c(basketless_nsb_down$gene, basketless_nsb_up$gene), "distinct", "shared"))

res_basket_features <- res_basket_features %>%
  mutate(distinct = ifelse(gene %in% c(basket_nsb_down$gene, basket_nsb_up$gene), "distinct", "shared"))

res_total_features <- res_total_features %>%
  mutate(distinct = ifelse(gene %in% c(total_nsb_down$gene, total_nsb_up$gene), "distinct", "shared"))
```

```{r EXPRESSION_by_full_set1, fig.height = 4}
plt1_exp_scatter <-  ggplot(res_basket_features, 
                            aes(x = log2FoldChange, y = mean_polya_vst, 
                                color = introns)) +
  labs(title = "Basket (n = 1379 + 1417)", x = "", y = "mean normalized poly(A) counts") +
  xlim(c(-6, 5)) + ylim(c(0, 18.5)) +
  geom_point(alpha = 1/5) +
  theme_minimal() +
  theme(legend.box="vertical") +
  scale_color_brewer(palette = "Set1", name = "intron", labels = c("does not contain an intron", "contains an intron")) 

plt2_exp_scatter <- ggplot(res_basketless_features,
                           aes(x = log2FoldChange, y = mean_polya_vst, 
                               color = introns)) +
  geom_point(alpha = 1/5) +
  labs(title = "Basketless (n = 484 + 603)", y = "", x = expression(log[2]~fold~change)) + 
  xlim(c(-6, 5)) + ylim(c(0, 18.5)) +
  theme_minimal() +
  theme(legend.box="vertical") +
  scale_color_brewer(palette = "Set1", name = "intron", labels = c("does not contain an intron", "contains an intron"))


# SWITCH FROM BASKET TO ALL PORE
plt3_exp_scatter <- ggplot(res_total_features, 
                           aes(x = log2FoldChange, y = mean_polya_vst, 
                               color = introns)) +
  labs(title = "All pore (n = 680 + 743)", y = "", x = "") +
  xlim(c(-6, 5)) + ylim(c(0, 18.5)) +
  geom_point(alpha = 1/5) +
  theme_minimal() +
  theme(legend.box="vertical") +
  scale_color_brewer(palette = "Set1", name = "intron", labels = c("does not contain an intron", "contains an intron")) 

ggarrange(plt1_exp_scatter, plt2_exp_scatter, plt3_exp_scatter,
          common.legend = T, ncol = 3, legend = "bottom")

```

```{r save_expression_by_full_set1_as_pdf, include = F, message=F}
pdf("publication_figures/expression_by_full_set1.pdf", width = 7.2, height = 3.5)
ggarrange(plt1_exp_scatter, plt2_exp_scatter, plt3_exp_scatter,
          common.legend = T, ncol = 3, legend = "bottom")
dev.off()

png("publication_figures/expression_by_full_set1.png", units = "in", res = 300, width = 7.2, height = 3.5)
ggarrange(plt1_exp_scatter, plt2_exp_scatter, plt3_exp_scatter,
          common.legend = T, ncol = 3, legend = "bottom")
dev.off()
```

## VST counts/expression density plots (Figure S7A)

```{r EXPRESSION_by_full_set2, fig.width = 10}
## Look at distinctly enriched in basket
basket_exp_density <- ggplot(res_basket_features, aes(x = mean_polya_vst, fill = change)) +
  geom_density(alpha = 0.5) +
  facet_wrap(~ifelse(introns == "no", "does not contain an intron", "contains an intron"),
             nrow = 2) +
  xlim(c(0, 18)) +
  theme_minimal() +
  theme(plot.title = element_text(size = 9),
        plot.title.position = "plot") +
  labs(title = "Basket (n = 1379 + 1417)", x = "") +
  scale_fill_discrete(name = expression(log[2]~fold~change), labels = c("negative", "positive"))

basket_exp_tbl <- res_basket_features %>%
  filter(gene %in% c(basket_nsb_up$gene, basket_nsb_down$gene)) %>%
  group_by(introns) %>% 
  do(tidy(t.test(mean_polya_vst ~ change, data = .)))  %>%
  dplyr::select(-estimate1, -estimate2, -parameter, -conf.low, -conf.high, -alternative) %>%
  ggtexttable(rows = NULL, theme = ttheme(base_size = 6))

basketless_exp_density <- ggplot(res_basketless_features, 
       aes(x = mean_polya_vst, fill = change)) +
  geom_density(alpha = .5) +
  theme_minimal() +
  facet_wrap(~ifelse(introns == "no", "does not contain an intron", "contains an intron"),
             nrow = 2) +
  xlim(c(0, 18)) +
  theme(plot.title = element_text(size = 9),
        plot.title.position = "plot") +
  labs(title = "Basketless (n = 484 + 603)", x = "mean normalized poly(A) counts", y="") +
  scale_fill_discrete(name = expression(log[2]~fold~change), labels = c("negative", "positive"))

basketless_exp_tbl <- res_basketless_features %>% 
  group_by(introns) %>% 
  do(tidy(t.test(mean_polya_vst ~ change, data = .))) %>%
  dplyr::select(-estimate1, -estimate2, -parameter, -conf.low, -conf.high, -alternative) %>%
  ggtexttable(rows = NULL, theme = ttheme(base_size = 6))

total_exp_density <- ggplot(res_total_features, 
       aes(x = mean_polya_vst, fill = change)) +
  geom_density(alpha = .5) +
  theme_minimal() +
  facet_wrap(~ifelse(introns == "no", "does not contain an intron", "contains an intron"),
             nrow = 2) +
  xlim(c(0, 18)) +
  theme(plot.title = element_text(size = 9),
        plot.title.position = "plot") +
  labs(title = "All pore (n = 680 + 743)", x = "", y = "") +
  scale_fill_discrete(name = expression(log[2]~fold~change), labels = c("negative", "positive"))

total_exp_tbl <- res_total_features %>% 
  group_by(introns) %>% 
  do(tidy(t.test(mean_polya_vst ~ change, data = .))) %>%
  dplyr::select(-estimate1, -estimate2, -parameter, -conf.low, -conf.high, -alternative) %>%
  ggtexttable(rows = NULL, theme = ttheme(base_size = 6))

ggarrange(basket_exp_density, basketless_exp_density, total_exp_density,
          basket_exp_tbl, basketless_exp_tbl, total_exp_tbl,
          common.legend = T, nrow = 2, ncol = 3, legend = "bottom", heights = c(3, 1))
```

```{r save_expression_by_full_set2_as_pdf, include = F, message=F}
pdf("publication_figures/expression_by_full_set2.pdf", width = 10, height = 5)
ggarrange(basket_exp_density, basketless_exp_density, total_exp_density,
          basket_exp_tbl, basketless_exp_tbl, total_exp_tbl,
          common.legend = T, nrow = 2, ncol = 3, legend = "bottom", heights = c(3, 1))
dev.off()

png("publication_figures/expression_by_full_set2.png", units = "in", res = 300, width = 10, height = 5)
ggarrange(basket_exp_density, basketless_exp_density, total_exp_density,
          basket_exp_tbl, basketless_exp_tbl, total_exp_tbl,
          common.legend = T, nrow = 2, ncol = 3, legend = "bottom", heights = c(3, 1))
dev.off()

```

## Length of transcripts & length of poly(A) tails (Figure 7D)

```{r polya_by_transcript_length}
plt5_polya <- ggplot(features %>%
                      filter(polya_length > 1), 
                    aes(y = width, x = polya_length)) +
  geom_point(alpha = .05) +
  theme_minimal() +
  theme(plot.title = element_text(size=9),
        plot.title.position = "plot") +
  xlim(c(0, 120)) + ylim(c(0,15000)) +
  labs(y = "transcript length", x = "", title = "whole transcriptome")


plt6_polya <- ggplot(res_basketless_features %>%
                filter(gene %in% c(total_basketless_basket_nsb_up$gene, 
                                   total_basketless_basket_nsb_down$gene)) %>%
                        filter(polya_length > 1), 
             aes(x = polya_length, y = width, color = change)) +
  geom_point(alpha = .2) +
  theme_minimal() +
  theme(plot.title = element_text(size=9),
        plot.title.position = "plot") +
  xlim(c(0, 120)) + ylim(c(0,15000)) +
  ggtitle(paste0("Basket, Basketless, All (n = ",
          total_basketless_basket_nsb_up_size, " + ",
          total_basketless_basket_nsb_down_size, ")")) +
  labs(y = "", x = "") +
  scale_color_discrete(name = expression(log[2]~fold~change), labels = c("negative", "positive"))

plt7_polya <- ggplot(res_basket_features %>%
                filter(gene %in% c(basket_nsb_up$gene, basket_nsb_down$gene))%>%
                        filter(polya_length > 1), 
              aes(x = polya_length, y = width, color = change)) +
  geom_point(alpha = .2) +
  xlim(c(0, 120)) + ylim(c(0,15000)) +
  theme_minimal() +
  theme(plot.title = element_text(size=9),
        plot.title.position = "plot") +
  ggtitle(paste0("Basket (n = ",
                 basket_nsb_up_size, " + ", 
                 basket_nsb_down_size, ")"))  +
  labs(y = "transcript length", x = "poly(A) tail length") +
  scale_color_discrete(name = expression(log[2]~fold~change), labels = c("negative", "positive"))

plt8_polya<- ggplot(res_basket_features %>%
                        filter(gene %in% c(total_basket_nsb_up$gene, 
                                           total_basket_nsb_down$gene)) %>%
                        filter(polya_length > 1), 
              aes(x = polya_length, y = width, color = change)) +
  geom_point(alpha = .2) +
  xlim(c(0, 120)) + ylim(c(0,15000)) +
  theme_minimal() +
  theme(plot.title = element_text(size=9),
        plot.title.position = "plot") +
  ggtitle(paste0("Basket, All (n = ", 
                 total_basket_nsb_up_size, " + ",
                 total_basket_nsb_down_size, ")")) +
  labs(y = "", x = "poly(A) tail length") +
  scale_color_discrete(name = expression(log[2]~fold~change), labels = c("negative", "positive"))

ggarrange(plt5_polya, plt6_polya, plt7_polya, plt8_polya,
          common.legend = T, legend = "bottom")
```

```{r save_polya_by_transcript_length_as_pdf, include = F, message = F}
pdf("publication_figures/polya_by_transcript_length.pdf", height = 4.5, width = 5)
ggarrange(plt5_polya, plt6_polya, plt7_polya, plt8_polya,
          common.legend = T, legend = "bottom")
dev.off()

png("publication_figures/polya_by_transcript_length.png", units = "in", res = 300, height = 4.5, width = 5)
ggarrange(plt5_polya, plt6_polya, plt7_polya, plt8_polya,
          common.legend = T, legend = "bottom")
dev.off()
```

## Length of transcripts & length of poly(A) tails (Figure S7B) 

```{r LENGTH_full_sets}
# wilcox.test() -- wilcoxon rank sum test. Is mean gene length different
#                  between two groups? (data are not normally distributed)

igg_wilcox_length <- wilcox.test(res_igg_features$width ~ res_igg_features$change) %>%
  broom::tidy() %>%
  mutate(set = "IgG ctrl")

gbp_wilcox_length <- wilcox.test(res_gbp_features$width ~ res_gbp_features$change) %>%
  broom::tidy() %>%
  mutate(set = "GFP nanobody ctrl")

basket_wilcox_length <- wilcox.test(res_basket_features$width ~ res_basket_features$change) %>%
  broom::tidy() %>%
  mutate(set = "basket")

basketless_wilcox_length <- wilcox.test(res_basketless_features$width ~ res_basketless_features$change) %>%
  broom::tidy() %>%
  mutate(set = "basketless")

total_wilcox_length <- wilcox.test(res_total_features$width ~ res_total_features$change) %>%
  broom::tidy() %>%
  mutate(set = "all pore")

all_wilcox_length <- rbind(igg_wilcox_length, 
                           gbp_wilcox_length,
                           basket_wilcox_length,
                           basketless_wilcox_length, 
                           total_wilcox_length) %>%
  as.data.frame() %>%
  mutate(method = "wilcox.test()") %>%
  dplyr::select(statistic, p.value, method, set)
  
plt1_length <- ggplot(res_basketless_features, aes(x = width)) +
  geom_density(data = features, aes(x = width), color ="black", linetype = "dashed", show.legend = F) +
  geom_density(alpha = .5, aes(fill = change)) +
  theme_minimal() +
  labs(title = "Basketless (n = 484 + 603)", x = "") + 
  #xlim(0, 15000) + ylim(0, 0.0011) +
  scale_fill_discrete(name = expression(log[2]~fold~change), labels = c("negative", "positive"))

plt2_length <- ggplot(res_total_features, aes(x = width)) +
  geom_density(alpha = .5, aes(fill = change)) +
  geom_density(data = features, aes(x = width), linetype = "dashed", color ="black", show.legend = F) +
  theme_minimal() +
  labs(title = "All pore (n = 680 + 743)", x = "")+ 
  xlim(0, 15000) +
  ylim(0, 0.0011) +
  scale_fill_discrete(name = expression(log[2]~fold~change), labels = c("negative", "positive"))

plt3_length <- ggplot(res_basket_features, aes(x = width)) +
  geom_density(alpha = .5, aes(fill = change)) +
  geom_density(data = features, aes(x = width), linetype = "dashed", color ="black", show.legend = F) +
  theme_minimal() +
  labs(title = "Basket (n = 1379 + 1417)", x = "transcript length")+ 
  xlim(0, 15000) +
  ylim(0, 0.0011)+
  scale_fill_discrete(name = expression(log[2]~fold~change), labels = c("negative", "positive"))

plt4_length <- ggplot(res_igg_features, aes(x = width)) +
  geom_density(alpha = .5, aes(fill = change)) +
  geom_density(data = features, aes(x = width), linetype = "dashed", color ="black", show.legend = F) +
  theme_minimal() +
  labs(title = "IgG control", y = "", x = "")+ 
  xlim(0, 15000) +
  ylim(0, 0.0011)+
  scale_fill_discrete(name = expression(log[2]~fold~change), labels = c("negative", "positive"))

plt5_length <- ggplot(res_gbp_features, aes(x = width)) +
  geom_density(alpha = .5, aes(fill = change)) +
  geom_density(data = features, aes(x = width), linetype = "dashed", color ="black", show.legend = F) +
  theme_minimal() +
  labs(title = "GFP nanobody control", y = "", x = "transcript length") + 
  xlim(0, 15000) +
  ylim(0, 0.0011)+
  scale_fill_discrete(name = expression(log[2]~fold~change), labels = c("negative", "positive"))

plt6_length <- ggtexttable(all_wilcox_length, rows = NULL, theme = ttheme(base_size = 7))

plt7_length <- ggplot(features, aes(x = width)) +
  theme_minimal() +
  geom_density() +
  labs(title = "transcriptome distribution", x = "transcript length")

ggarrange1 <- ggarrange(plt1_length, plt2_length, plt3_length, 
                        nrow = 3, legend = "none", common.legend = T)
ggarrange2 <- ggarrange(plt4_length, plt5_length, plt6_length,
                        common.legend = T, nrow = 3, legend = "bottom")

ggarrange(ggarrange1, ggarrange2, ncol = 2, common.legend = T)
```

```{r save_length_as_pdf, include = F, message = F}
pdf("publication_figures/transcript_length.pdf", height = 6, width = 6)
ggarrange(ggarrange1, ggarrange2, ncol = 2, common.legend = T)
dev.off()

png("publication_figures/transcript_length.png", units = "in", res = 300, height = 6, width = 6)
ggarrange(ggarrange1, ggarrange2, ncol = 2, common.legend = T)
dev.off()
```

```{r POLYA_full_set, fig.height=3}
# wilcox.test() -- wilcoxon rank sum test. Is mean polya tail lengths different
#                  between two groups? (data are not normally distributed)

# all_basket_basketless_wilcox_polya2 <- res_basketless_features %>%
#   filter(gene %in% c(total_basketless_basket_nsb_up$gene, 
#                      total_basketless_basket_nsb_down$gene)) %>%
#   filter(polya_length > 1) %>%
#   do(tidy(wilcox.test(polya_length ~ change, data = .))) %>%
#   mutate(set = "basket, basketless, all")
# 
# basket_wilcox_polya2 <- res_basket_features %>%
#   filter(gene %in% c(basket_nsb_up$gene, basket_nsb_down$gene)) %>%
#   filter(polya_length > 1) %>%
#   do(tidy(wilcox.test(polya_length ~ change, data = .))) %>%
#   mutate(set = "basket")
# 
# basket_all_wilcox_polya2 <- res_basket_features %>%
#   filter(gene %in% c(total_basket_nsb_up$gene, 
#                      total_basket_nsb_down$gene)) %>%
#   filter(polya_length > 1) %>%
#   do(tidy(wilcox.test(polya_length ~ change, data = .))) %>%
#   mutate(set = "basket, all")
# 
# all_wilcox_polya2 <- rbind(basket_wilcox_polya2,
#                            basket_all_wilcox_polya2,
#                            all_basket_basketless_wilcox_polya2) %>%
#   as.data.frame() %>%
#   mutate(method = "wilcox.test()") %>%
#   dplyr::select(statistic, p.value, method, set)

plt1_polya <- ggplot(res_basketless_features %>% filter(polya_length > 1), 
                     aes(x = polya_length)) +
  geom_density(alpha = .5, aes(fill = change)) +
  geom_density(data = features %>% filter(polya_length > 1), aes(x = polya_length), linetype = "dashed", color = "black", show.legend = F) +
  ylim(c(0, 0.09)) +
  theme_minimal() +
  theme(plot.title = element_text(size=9),
        plot.title.position = "plot") +
  labs(title = "Basketless (n = 484 + 603)", y = "", x = "poly(A) tail length") +
  scale_fill_discrete(name = expression(log[2]~fold~change), labels = c("negative", "positive"))

plt2_polya <- ggplot(res_basket_features %>% filter(polya_length > 1), 
              aes(x = polya_length)) +
  geom_density(alpha = .5, aes(fill = change)) +
  geom_density(data = features %>% filter(polya_length > 1), aes(x = polya_length), linetype = "dashed", color = "black", show.legend = F) +
  ylim(c(0, 0.09)) +
  theme_minimal() +
  theme(plot.title = element_text(size=9),
        plot.title.position = "plot") +
  labs(title = "Basket (n = 1379 + 1417)", x = "") + 
  scale_fill_discrete(name = expression(log[2]~fold~change), labels = c("negative", "positive"))

plt3_polya <- ggplot(res_total_features %>% filter(polya_length > 1), 
              aes(x = polya_length)) +
  geom_density(alpha = .5, aes(fill = change)) +
  geom_density(data = features %>% filter(polya_length > 1), aes(x = polya_length), linetype = "dashed", color = "black", show.legend = F) +
  ylim(c(0, 0.09)) +
  theme_minimal() +
  theme(plot.title = element_text(size=9),
        plot.title.position = "plot") +
  labs(title = "All pore (n = 680 + 743)", y = "", x = "") +
  scale_fill_discrete(name = expression(log[2]~fold~change), labels = c("negative", "positive"))


ggarrange(plt2_polya, plt1_polya, plt3_polya,
          legend = "bottom", common.legend = T, ncol = 3)
```

```{r save_polya_as_pdf, include = F, message = F}
pdf("publication_figures/polya_length.pdf", height = 2.5, width = 5)
ggarrange(plt2_polya, plt1_polya, plt3_polya,
          legend = "bottom", common.legend = T, ncol = 3)
dev.off()

png("publication_figures/polya_length.png", units = "in", res = 300, height = 2.5, width = 5)
ggarrange(plt2_polya, plt1_polya, plt3_polya,
          legend = "bottom", common.legend = T, ncol = 3)
dev.off()
```


## Regression models

### basket

```{r regression_models}
res_basket_features$introns2 <- ifelse(res_basket_features$introns == "yes", 1, 0)
basket_width_intron <- lm(formula = log2FoldChange ~ width + introns2, data = res_basket_features)
basket_width_intron <- lm(formula = log2FoldChange ~ width + introns2 + polya_length + mean_polya_vst, data = res_basket_features %>% filter(polya_length > 1), na.action = "na.omit")
basket_width_intron %>% glance() %>% kable() %>% kable_styling()
basket_width_intron %>% tidy() %>% kable() %>% kable_styling()
```

### basketless

```{r}
res_basketless_features$introns2 <- ifelse(res_basketless_features$introns == "yes", 1, 0)
basketless_width_intron <- lm(formula = log2FoldChange ~ width + introns2 + polya_length + mean_polya_vst, data = res_basketless_features %>% filter(polya_length > 1), na.action = "na.omit")
basketless_width_intron %>% glance() %>% kable() %>% kable_styling()
basketless_width_intron %>% tidy() %>% kable() %>% kable_styling()
```

### all pore
```{r}
res_total_features$introns2 <- ifelse(res_total_features$introns == "yes", 1, 0)
total_width_intron <- lm(formula = log2FoldChange ~ width + introns2 + polya_length + mean_polya_vst, data = res_total_features)
total_width_intron %>% glance() %>% kable() %>% kable_styling()
total_width_intron %>% tidy() %>% kable() %>% kable_styling()
```

## Alternatives -- histograms instead of density plots

### VST counts/expression density plots (Figure S7A)

```{r EXPRESSION_by_full_set22, fig.width = 10}
## Look at distinctly enriched in basket
basket_exp_histogram <- ggplot() +
  geom_histogram(data = features, aes(x = mean_polya_vst), alpha = .5, color = "grey", show.legend = F) +
  geom_histogram(data = res_basket_features, aes(x = mean_polya_vst, fill = change), alpha = 0.5) +
  facet_wrap(~ifelse(introns == "no", "does not contain an intron", "contains an intron"),
             nrow = 2) +
  theme_minimal() +
  theme(plot.title = element_text(size = 9),
        plot.title.position = "plot") +
  labs(title = "Basket (n = 1379 + 1417)", x = "") +
  scale_fill_discrete(name = expression(log[2]~fold~change), labels = c("negative", "positive"))

basket_exp_tbl <- res_basket_features %>%
  filter(gene %in% c(basket_nsb_up$gene, basket_nsb_down$gene)) %>%
  group_by(introns) %>% 
  do(tidy(t.test(mean_polya_vst ~ change, data = .)))  %>%
  dplyr::select(-estimate1, -estimate2, -parameter, -conf.low, -conf.high, -alternative) %>%
  ggtexttable(rows = NULL, theme = ttheme(base_size = 6))

basketless_exp_histogram <- ggplot() +
  geom_histogram(data = features, aes(x = mean_polya_vst), alpha = .5, color = "grey", show.legend = F) +
  geom_histogram(data = res_basketless_features, aes(x = mean_polya_vst, fill = change), alpha = 0.5) +
  facet_wrap(~ifelse(introns == "no", "does not contain an intron", "contains an intron"),
             nrow = 2) +
  theme_minimal() +
  theme(plot.title = element_text(size = 9),
        plot.title.position = "plot") +
  labs(title = "Basketless (n = 484 + 603)", x = "mean normalized poly(A) counts", y="") +
  scale_fill_discrete(name = expression(log[2]~fold~change), labels = c("negative", "positive"))

basketless_exp_tbl <- res_basketless_features %>% 
  group_by(introns) %>% 
  do(tidy(t.test(mean_polya_vst ~ change, data = .))) %>%
  dplyr::select(-estimate1, -estimate2, -parameter, -conf.low, -conf.high, -alternative) %>%
  ggtexttable(rows = NULL, theme = ttheme(base_size = 6))

total_exp_histogram <- ggplot() +
  geom_histogram(data = features, aes(x = mean_polya_vst), alpha = .5, color = "grey", show.legend = F) +
  geom_histogram(data = res_total_features, aes(x = mean_polya_vst, fill = change), alpha = 0.5) +
  facet_wrap(~ifelse(introns == "no", "does not contain an intron", "contains an intron"),
             nrow = 2) +
  theme_minimal() +
  theme(plot.title = element_text(size = 9),
        plot.title.position = "plot") +
  labs(title = "All pore (n = 680 + 743)", x = "", y = "") +
  scale_fill_discrete(name = expression(log[2]~fold~change), labels = c("negative", "positive"))

total_exp_tbl <- res_total_features %>% 
  group_by(introns) %>% 
  do(tidy(t.test(mean_polya_vst ~ change, data = .))) %>%
  dplyr::select(-estimate1, -estimate2, -parameter, -conf.low, -conf.high, -alternative) %>%
  ggtexttable(rows = NULL, theme = ttheme(base_size = 6))

ggarrange(basket_exp_histogram, basketless_exp_histogram, total_exp_histogram,
          basket_exp_tbl, basketless_exp_tbl, total_exp_tbl,
          common.legend = T, nrow = 2, ncol = 3, legend = "bottom", heights = c(3, 1))
```

```{r save_expression_by_full_set2_as_pdf2, include = F, message=F}
pdf("publication_figures/expression_by_full_set2_alt.pdf", width = 10, height = 5)
ggarrange(basket_exp_histogram, basketless_exp_histogram, total_exp_histogram,
          basket_exp_tbl, basketless_exp_tbl, total_exp_tbl,
          common.legend = T, nrow = 2, ncol = 3, legend = "bottom", heights = c(3, 1))
dev.off()

png("publication_figures/expression_by_full_set2_alt.png", units = "in", res = 300, width = 10, height = 5)
ggarrange(basket_exp_histogram, basketless_exp_histogram, total_exp_histogram,
          basket_exp_tbl, basketless_exp_tbl, total_exp_tbl,
          common.legend = T, nrow = 2, ncol = 3, legend = "bottom", heights = c(3, 1))
dev.off()

```


### Length of transcripts & length of poly(A) tails (Figure S7B) 

```{r LENGTH_full_sets2}
# wilcox.test() -- wilcoxon rank sum test. Is mean gene length different
#                  between two groups? (data are not normally distributed)

igg_wilcox_length <- wilcox.test(res_igg_features$width ~ res_igg_features$change) %>%
  broom::tidy() %>%
  mutate(set = "IgG ctrl")

gbp_wilcox_length <- wilcox.test(res_gbp_features$width ~ res_gbp_features$change) %>%
  broom::tidy() %>%
  mutate(set = "GFP nanobody ctrl")

basket_wilcox_length <- wilcox.test(res_basket_features$width ~ res_basket_features$change) %>%
  broom::tidy() %>%
  mutate(set = "basket")

basketless_wilcox_length <- wilcox.test(res_basketless_features$width ~ res_basketless_features$change) %>%
  broom::tidy() %>%
  mutate(set = "basketless")

total_wilcox_length <- wilcox.test(res_total_features$width ~ res_total_features$change) %>%
  broom::tidy() %>%
  mutate(set = "all pore")

all_wilcox_length <- rbind(igg_wilcox_length, 
                           gbp_wilcox_length,
                           basket_wilcox_length,
                           basketless_wilcox_length, 
                           total_wilcox_length) %>%
  as.data.frame() %>%
  mutate(method = "wilcox.test()") %>%
  dplyr::select(statistic, p.value, method, set)
  
plt1_length <- ggplot(res_basketless_features, aes(x = width)) +
  geom_histogram(data = features, aes(x = width), color ="grey", show.legend = F, alpha = .5) +
  geom_histogram(alpha = .5, aes(fill = change)) +
  theme_minimal() +
  labs(title = "Basketless (n = 484 + 603)", x = "") + 
  #xlim(0, 15000) + ylim(0, 0.0011) +
  scale_fill_discrete(name = expression(log[2]~fold~change), labels = c("negative", "positive"))

plt2_length <- ggplot(res_total_features, aes(x = width)) +
  geom_histogram(data = features, aes(x = width), color ="grey", show.legend = F, alpha = .5) +
  geom_histogram(alpha = .5, aes(fill = change)) +
  theme_minimal() +
  labs(title = "All pore (n = 680 + 743)", x = "")+ 
  #xlim(0, 15000) +
  #ylim(0, 0.0011) +
  scale_fill_discrete(name = expression(log[2]~fold~change), labels = c("negative", "positive"))

plt3_length <- ggplot(res_basket_features, aes(x = width)) +
  geom_histogram(data = features, aes(x = width), color ="grey", show.legend = F, alpha = .5) +
  geom_histogram(alpha = .5, aes(fill = change)) +
  theme_minimal() +
  labs(title = "Basket (n = 1379 + 1417)", x = "transcript length")+ 
  # xlim(0, 15000) +
  # ylim(0, 0.0011)+
  scale_fill_discrete(name = expression(log[2]~fold~change), labels = c("negative", "positive"))

plt4_length <- ggplot(res_igg_features, aes(x = width)) +
  geom_histogram(data = features, aes(x = width), color ="grey", show.legend = F, alpha = .5) +
  geom_histogram(alpha = .5, aes(fill = change)) +
  theme_minimal() +
  labs(title = "IgG control", y = "", x = "")+ 
  # xlim(0, 15000) +
  # ylim(0, 0.0011)+
  scale_fill_discrete(name = expression(log[2]~fold~change), labels = c("negative", "positive"))

plt5_length <- ggplot(res_gbp_features, aes(x = width)) +
  geom_histogram(data = features, aes(x = width), color ="grey", show.legend = F, alpha = .5) +
  geom_histogram(alpha = .5, aes(fill = change)) +
  theme_minimal() +
  labs(title = "GFP nanobody control", y = "", x = "transcript length") + 
  # xlim(0, 15000) +
  # ylim(0, 0.0011)+
  scale_fill_discrete(name = expression(log[2]~fold~change), labels = c("negative", "positive"))

plt6_length <- ggtexttable(all_wilcox_length, rows = NULL, theme = ttheme(base_size = 7))

ggarrange1 <- ggarrange(plt1_length, plt2_length, plt3_length, 
                        nrow = 3, legend = "none", common.legend = T)
ggarrange2 <- ggarrange(plt4_length, plt5_length, plt6_length,
                        common.legend = T, nrow = 3, legend = "bottom")

ggarrange(ggarrange1, ggarrange2, ncol = 2, common.legend = T)
```

```{r save_length_as_pdf2, include = F, message = F}
pdf("publication_figures/transcript_length_alt.pdf", height = 6, width = 6)
ggarrange(ggarrange1, ggarrange2, ncol = 2, common.legend = T)
dev.off()

png("publication_figures/transcript_length_alt.png", units = "in", res = 300, height = 6, width = 6)
ggarrange(ggarrange1, ggarrange2, ncol = 2, common.legend = T)
dev.off()
```

```{r POLYA_full_set2, fig.height=3}
plt1_polya <- ggplot(res_basketless_features %>% filter(polya_length > 1), 
                     aes(x = polya_length)) +
  geom_histogram(data = features %>% filter(polya_length > 1), aes(x = polya_length), alpha = .5, color = "grey", show.legend = F) +
  geom_histogram(alpha = .5, aes(fill = change)) +
  theme_minimal() +
  theme(plot.title = element_text(size=9),
        plot.title.position = "plot") +
  labs(title = "Basketless (n = 484 + 603)", y = "", x = "poly(A) tail length") +
  scale_fill_discrete(name = expression(log[2]~fold~change), labels = c("negative", "positive")) 
plt2_polya <- ggplot(res_basket_features %>% filter(polya_length > 1), 
              aes(x = polya_length)) +
  geom_histogram(data = features %>% filter(polya_length > 1), aes(x = polya_length), alpha = .5, color = "grey", show.legend = F) +
  geom_histogram(alpha = .5, aes(fill = change)) +
  theme_minimal() +
  theme(plot.title = element_text(size=9),
        plot.title.position = "plot") +
  labs(title = "Basket (n = 1379 + 1417)", x = "") + 
  scale_fill_discrete(name = expression(log[2]~fold~change), labels = c("negative", "positive"))

plt3_polya <- ggplot(res_total_features %>% filter(polya_length > 1), 
              aes(x = polya_length)) +
  geom_histogram(data = features %>% filter(polya_length > 1), aes(x = polya_length), alpha = .5, color = "grey", show.legend = F) +
  geom_histogram(alpha = .5, aes(fill = change)) +
  theme_minimal() +
  theme(plot.title = element_text(size=9),
        plot.title.position = "plot") +
  labs(title = "All pore (n = 680 + 743)", y = "", x = "") +
  scale_fill_discrete(name = expression(log[2]~fold~change), labels = c("negative", "positive"))


ggarrange(plt2_polya, plt1_polya, plt3_polya,
          legend = "bottom", common.legend = T, ncol = 3)
```

```{r save_polya_as_pdf2, include = F, message = F}
pdf("publication_figures/polya_length_alt.pdf", height = 2.5, width = 5)
ggarrange(plt2_polya, plt1_polya, plt3_polya,
          legend = "bottom", common.legend = T, ncol = 3)
dev.off()

png("publication_figures/polya_length_alt.png", units = "in", res = 300, height = 2.5, width = 5)
ggarrange(plt2_polya, plt1_polya, plt3_polya,
          legend = "bottom", common.legend = T, ncol = 3)
dev.off()
```

